<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-GB" xml:lang="en-GB">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="James A. Fellows Yates" />
  <meta name="author" content="Thiseas C. Lamnidis" />
  <meta name="author" content="Maxime Borry" />
  <meta name="author" content="Aida Andrades Valtueña" />
  <meta name="author" content="Zandra Fagernäs" />
  <meta name="author" content="Stephen Clayton" />
  <meta name="author" content="Maxime U. Garcia" />
  <meta name="author" content="Judith Neukamm" />
  <meta name="author" content="Alexander Peltzer" />
  <meta name="dcterms.date" content="2020-06-05" />
  <meta name="keywords" content="nf-core, nextflow, ancientDNA, aDNA, palaeogenetics, archaeogenetics, bioinformatics" />
  <title>Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      {   }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span.al { color: #ff0000; font-weight: bold; } /* Alert */
    code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
    code span.at { color: #7d9029; } /* Attribute */
    code span.bn { color: #40a070; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #4070a0; } /* Char */
    code span.cn { color: #880000; } /* Constant */
    code span.co { color: #60a0b0; font-style: italic; } /* Comment */
    code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
    code span.do { color: #ba2121; font-style: italic; } /* Documentation */
    code span.dt { color: #902000; } /* DataType */
    code span.dv { color: #40a070; } /* DecVal */
    code span.er { color: #ff0000; font-weight: bold; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #40a070; } /* Float */
    code span.fu { color: #06287e; } /* Function */
    code span.im { } /* Import */
    code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
    code span.kw { color: #007020; font-weight: bold; } /* Keyword */
    code span.op { color: #666666; } /* Operator */
    code span.ot { color: #007020; } /* Other */
    code span.pp { color: #bc7a00; } /* Preprocessor */
    code span.sc { color: #4070a0; } /* SpecialChar */
    code span.ss { color: #bb6688; } /* SpecialString */
    code span.st { color: #4070a0; } /* String */
    code span.va { color: #19177c; } /* Variable */
    code span.vs { color: #4070a0; } /* VerbatimString */
    code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager" />
  <meta name="citation_title" content="Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager" />
  <meta property="og:title" content="Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager" />
  <meta property="twitter:title" content="Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager" />
  <meta name="dc.date" content="2020-06-05" />
  <meta name="citation_publication_date" content="2020-06-05" />
  <meta name="dc.language" content="en-GB" />
  <meta name="citation_language" content="en-GB" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="James A. Fellows Yates" />
  <meta name="citation_author_institution" content="Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany" />
  <meta name="citation_author_orcid" content="0000-0001-5585-6277" />
  <meta name="twitter:creator" content="@jfy133" />
  <meta name="citation_author" content="Thiseas C. Lamnidis" />
  <meta name="citation_author_institution" content="Population Genetics Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany" />
  <meta name="citation_author_orcid" content="0000-0003-4485-8570" />
  <meta name="twitter:creator" content="@TCLamnidis" />
  <meta name="citation_author" content="Maxime Borry" />
  <meta name="citation_author_institution" content="Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany" />
  <meta name="citation_author_orcid" content="0000-0001-9140-7559" />
  <meta name="twitter:creator" content="@notmaxib" />
  <meta name="citation_author" content="Aida Andrades Valtueña" />
  <meta name="citation_author_institution" content="Computational Pathogenomics Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany" />
  <meta name="citation_author_orcid" content="0000-0002-1737-2228" />
  <meta name="twitter:creator" content="@aidaanva" />
  <meta name="citation_author" content="Zandra Fagernäs" />
  <meta name="citation_author_institution" content="Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany" />
  <meta name="citation_author_orcid" content="0000-0003-2667-3556" />
  <meta name="twitter:creator" content="@ZandraSelina" />
  <meta name="citation_author" content="Stephen Clayton" />
  <meta name="citation_author_institution" content="Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany" />
  <meta name="citation_author_orcid" content="0000-0001-5223-9695" />
  <meta name="citation_author" content="Maxime U. Garcia" />
  <meta name="citation_author_institution" content="Department of Oncology-Pathology, Karolinska Institutet, Stockholm, Sweden" />
  <meta name="citation_author_orcid" content="0000-0003-2827-9261" />
  <meta name="twitter:creator" content="@gau" />
  <meta name="citation_author" content="Judith Neukamm" />
  <meta name="citation_author_institution" content="Palaeogenetics Group, Institute of Evolutionary Medicine, University of Zurich, Zürich, Switzerland" />
  <meta name="citation_author_orcid" content="0000-0001-8141-566X" />
  <meta name="twitter:creator" content="@JudithNeukamm" />
  <meta name="citation_author" content="Alexander Peltzer" />
  <meta name="citation_author_institution" content="Quantitative Biology Center (QBiC), Eberhard-Karls-Universität, Tübingen, Germany" />
  <meta name="citation_author_orcid" content="0000-0002-6503-2180" />
  <meta name="twitter:creator" content="@alex_peltzer" />
  <link rel="canonical" href="https://apeltzer.github.io/eager2-paper/" />
  <meta property="og:url" content="https://apeltzer.github.io/eager2-paper/" />
  <meta property="twitter:url" content="https://apeltzer.github.io/eager2-paper/" />
  <meta name="citation_fulltext_html_url" content="https://apeltzer.github.io/eager2-paper/" />
  <meta name="citation_pdf_url" content="https://apeltzer.github.io/eager2-paper/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://apeltzer.github.io/eager2-paper/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://apeltzer.github.io/eager2-paper/v/8b6289d9495312866561826fc24c26addc278af5/" />
  <meta name="manubot_html_url_versioned" content="https://apeltzer.github.io/eager2-paper/v/8b6289d9495312866561826fc24c26addc278af5/" />
  <meta name="manubot_pdf_url_versioned" content="https://apeltzer.github.io/eager2-paper/v/8b6289d9495312866561826fc24c26addc278af5/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <meta property="og:image" content="https://github.com/apeltzer/eager2-paper/raw/8b6289d9495312866561826fc24c26addc278af5/content/images/nf-core-eager_social_preview.png" />
  <meta property="twitter:image" content="https://github.com/apeltzer/eager2-paper/raw/8b6289d9495312866561826fc24c26addc278af5/content/images/nf-core-eager_social_preview.png" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager</h1>
</header>
<p><small><em>
This manuscript
(<a href="https://apeltzer.github.io/eager2-paper/v/8b6289d9495312866561826fc24c26addc278af5/">permalink</a>)
was automatically generated
from <a href="https://github.com/apeltzer/eager2-paper/tree/8b6289d9495312866561826fc24c26addc278af5">apeltzer/eager2-paper@8b6289d</a>
on June 5, 2020.
</em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>James A. Fellows Yates</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-5585-6277">0000-0001-5585-6277</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/jfy133">jfy133</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/jfy133">jfy133</a><br>
<small>
Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
· Funded by Max Planck Society
</small></p></li>
<li><p><strong>Thiseas C. Lamnidis</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4485-8570">0000-0003-4485-8570</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/TCLamnidis">TCLamnidis</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/TCLamnidis">TCLamnidis</a><br>
<small>
Population Genetics Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
· Funded by Max Planck Society
</small></p></li>
<li><p><strong>Maxime Borry</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-9140-7559">0000-0001-9140-7559</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/maxibor">maxibor</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/notmaxib">notmaxib</a><br>
<small>
Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
· Funded by Max Planck Society
</small></p></li>
<li><p><strong>Aida Andrades Valtueña</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-1737-2228">0000-0002-1737-2228</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/aidaanva">aidaanva</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/aidaanva">aidaanva</a><br>
<small>
Computational Pathogenomics Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
· Funded by Max Planck Society
</small></p></li>
<li><p><strong>Zandra Fagernäs</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-2667-3556">0000-0003-2667-3556</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/ZandraFagernas">ZandraFagernas</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/ZandraSelina">ZandraSelina</a><br>
<small>
Microbiome Sciences Group, Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
· Funded by Max Planck Society
</small></p></li>
<li><p><strong>Stephen Clayton</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-5223-9695">0000-0001-5223-9695</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/sc13-bioinf">sc13-bioinf</a><br>
<small>
Department of Archaeogenetics, Max-Planck-Institute for the Science of Human History, Jena, Germany
· Funded by Max Planck Society
</small></p></li>
<li><p><strong>Maxime U. Garcia</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-2827-9261">0000-0003-2827-9261</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/MaxUlysse">MaxUlysse</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/gau">gau</a><br>
<small>
Department of Oncology-Pathology, Karolinska Institutet, Stockholm, Sweden
· Funded by Barncancerfonden
</small></p></li>
<li><p><strong>Judith Neukamm</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-8141-566X">0000-0001-8141-566X</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/JudithNeukamm">JudithNeukamm</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/JudithNeukamm">JudithNeukamm</a><br>
<small>
Palaeogenetics Group, Institute of Evolutionary Medicine, University of Zurich, Zürich, Switzerland
</small></p></li>
<li><p><strong>Alexander Peltzer</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-6503-2180">0000-0002-6503-2180</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/apeltzer">apeltzer</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/alex_peltzer">alex_peltzer</a><br>
<small>
Quantitative Biology Center (QBiC), Eberhard-Karls-Universität, Tübingen, Germany
</small></p></li>
</ul>
<h2 class="page_break_before" id="abstract">Abstract</h2>
<p>The broadening utilization of ancient DNA (aDNA) to address archaeological,
palaeontological and biological questions is resulting in a rising diversity in
the size of laboratories and scale of analyses being performed. In the context
of this heterogeneous landscape, we present nf-core/eager, an advanced and
entirely redesigned pipeline for the analysis of ancient DNA genomic data.
nf-core/eager builds on existing ideas and concepts introduced in the original
EAGER pipeline, and improves various aspects of the analysis procedure by
building on computational frameworks such as Nextflow and nf-core. The pipeline
aims to address three main points: accessibility and adaptability to different
research groups and their computing configurations, reproducibility to ensure
robust analytical standards in the field, and updating the EAGER pipeline to the
latest routine ancient genomic practises. This new version of EAGER has been
developed within the nf-core initiative, to ensure high quality software
development and maintenance support; contributing to a long-term lifecycle for
the pipeline. nf-core/eager will assist in ensuring that ancient DNA sequencing
data can be utilised by a diverse range of research groups and fields.</p>
<h2 id="introduction">Introduction</h2>
<p>Ancient DNA (aDNA) has become a widely accepted source of biological data,
helping to provide new perspective for a range of fields including archaeology,
ecology, cultural heritage, and palaeontology. The utilisation of
next-generation-sequencing has allowed the recovery of aDNA from a wide variety
of sources, including but not limited to, the skeletal remains of animals
<span class="citation" data-cites="86EfjHO2 1EJZxEzq1 7nLP30p0 ZMf8LvSn">[<a href="#ref-86EfjHO2" role="doc-biblioref">1</a>,<a href="#ref-1EJZxEzq1" role="doc-biblioref">2</a>,<a href="#ref-7nLP30p0" role="doc-biblioref">3</a>,<a href="#ref-ZMf8LvSn" role="doc-biblioref">4</a>]</span>, modern and archaic
humans <span class="citation" data-cites="tAnEwrSn dHnkKsKe qngWumct">[<a href="#ref-tAnEwrSn" role="doc-biblioref">5</a>,<a href="#ref-dHnkKsKe" role="doc-biblioref">6</a>,<a href="#ref-qngWumct" role="doc-biblioref">7</a>]</span>, bacteria <span class="citation" data-cites="oasyTY16 1APjJqbjj 4YuC8sCQ">[<a href="#ref-oasyTY16" role="doc-biblioref">8</a>,<a href="#ref-1APjJqbjj" role="doc-biblioref">9</a>,<a href="#ref-4YuC8sCQ" role="doc-biblioref">10</a>]</span>, viruses
<span class="citation" data-cites="CNvvxsK7 1HEgHFzOs">[<a href="#ref-CNvvxsK7" role="doc-biblioref">11</a>,<a href="#ref-1HEgHFzOs" role="doc-biblioref">12</a>]</span>, plants
<span class="citation" data-cites="1001q664T 6MKvilVe">[<a href="#ref-1001q664T" role="doc-biblioref">13</a>,<a href="#ref-6MKvilVe" role="doc-biblioref">14</a>]</span>, coprolites
<span class="citation" data-cites="fK2md3DI h6xmK5cY">[<a href="#ref-fK2md3DI" role="doc-biblioref">15</a>,<a href="#ref-h6xmK5cY" role="doc-biblioref">16</a>]</span>, dental calculus
<span class="citation" data-cites="10wcktDCE s3l2F2ef">[<a href="#ref-10wcktDCE" role="doc-biblioref">17</a>,<a href="#ref-s3l2F2ef" role="doc-biblioref">18</a>]</span>, sediments
<span class="citation" data-cites="YBHwLDej 64ZHuA4N">[<a href="#ref-YBHwLDej" role="doc-biblioref">19</a>,<a href="#ref-64ZHuA4N" role="doc-biblioref">20</a>]</span>, medical slides
<span class="citation" data-cites="WIkaempF">[<a href="#ref-WIkaempF" role="doc-biblioref">21</a>]</span>, parchment <span class="citation" data-cites="M0SKmPyu">[<a href="#ref-M0SKmPyu" role="doc-biblioref">22</a>]</span>, and most
recently ancient ‘chewing gum’ <span class="citation" data-cites="NXcxgeob 10155v3VM">[<a href="#ref-NXcxgeob" role="doc-biblioref">23</a>,<a href="#ref-10155v3VM" role="doc-biblioref">24</a>]</span>. Improvement in laboratory protocols to increase
yields of otherwise trace amount of DNA has at the same time led to studies that
can total hundreds of ancient individuals <span class="citation" data-cites="aSy3Srju wT5SMjro">[<a href="#ref-aSy3Srju" role="doc-biblioref">25</a>,<a href="#ref-wT5SMjro" role="doc-biblioref">26</a>]</span>, spanning single <span class="citation" data-cites="1AkUQBPuG">[<a href="#ref-1AkUQBPuG" role="doc-biblioref">27</a>]</span> to
thousands of organisms <span class="citation" data-cites="10wcktDCE">[<a href="#ref-10wcktDCE" role="doc-biblioref">17</a>]</span>. These differences of disciplines
have led to a heterogeneous landscape in terms of types of analyses, and thus
various types of computing resources required by different labs. Particularly,
this is true for regions of the world where aDNA as a field of research is still
emerging <span class="citation" data-cites="TwlGYuW7 h1qNDSRd">[<a href="#ref-TwlGYuW7" role="doc-biblioref">28</a>,<a href="#ref-h1qNDSRd" role="doc-biblioref">29</a>]</span>.</p>
<p>The degraded nature of aDNA poses an extra layer of complexity to standard
modern genomic analysis. Through a variety of processes <span class="citation" data-cites="Zp5bQCLJ">[<a href="#ref-Zp5bQCLJ" role="doc-biblioref">30</a>]</span>,
DNA molecules will fragment overtime resulting in ultra-short molecules
<span class="citation" data-cites="ZeQp0m3j">[<a href="#ref-ZeQp0m3j" role="doc-biblioref">31</a>]</span>). These sequences have low nucleotide complexity
making it difficult to identify with precision which part of the genome a read
is derived from. Furthermore, when fragmentation is not a ‘clean break’ this can
lead to uneven ends with single-stranded ‘overhangs’ at end of molecules, which
are susceptible to chemical processes such as deamination that lead to
mis-incorporation of bases during library construction
<span class="citation" data-cites="dbfFgAgr">[<a href="#ref-dbfFgAgr" role="doc-biblioref">32</a>]</span>. On top of this, taphonomic processes such as
heat, moisture, and microbial and burial environment processes lead to varying
rates of degradation <span class="citation" data-cites="NVqMN9QU jz1XHWKM">[<a href="#ref-NVqMN9QU" role="doc-biblioref">33</a>,<a href="#ref-jz1XHWKM" role="doc-biblioref">34</a>]</span>, where the original DNA content of a
sample is lost and supplanted by modern environmental DNA. Later handling by
archaeologists, museum curators, and scientists can also contribute ‘modern’
contamination. While these characteristics can help provide evidence towards the
‘authenticity’ of true aDNA sequences (e.g. aDNA C&gt;T ‘damage’ profiles
<span class="citation" data-cites="14YGEdSwc">[<a href="#ref-14YGEdSwc" role="doc-biblioref">35</a>]</span>), they also pose specific challenges such
as unspecific DNA alignment and/or low coverage and miscoding lesions resulting
in low-confidence genotyping. These factors often lead to prohibitive sequencing
costs to retrieve enough data for modern NGS data pipelines (e.g. &gt; 1 billion
reads for a 1X depth coverage <em>Yersinia pestis</em> genome
<span class="citation" data-cites="1SX8gKlw">[<a href="#ref-1SX8gKlw" role="doc-biblioref">36</a>]</span>), and thus require aDNA tailored methods and
techniques to overcome these challenges.</p>
<p>Two previously published and commonly used pipelines in the field are PALEOMIX
<span class="citation" data-cites="1H3banOxK">[<a href="#ref-1H3banOxK" role="doc-biblioref">37</a>]</span> and EAGER <span class="citation" data-cites="12TkhsLGD">[<a href="#ref-12TkhsLGD" role="doc-biblioref">38</a>]</span>. These
two pipelines take a similar approach to link together standard tools used for
Illumina NGS data processing (sequencing quality control, sequencing adapter
removal/and or paired-end read merging, mapping of reads to a reference genome,
genotyping, etc.), but with a specific focus on tools that are designed for or
well-suited for aDNA (such as bwa aln that works well on ultra-short molecules
<span class="citation" data-cites="roi5bwXL">[<a href="#ref-roi5bwXL" role="doc-biblioref">39</a>]</span> and mapDamage
<span class="citation" data-cites="Rhiz5BNL">[<a href="#ref-Rhiz5BNL" role="doc-biblioref">40</a>]</span> for aDNA characteristics evaluation). Yet,
neither of these pipelines have had major updates to bring them in-line with
current routine aDNA analyses. Metagenomic screening of off-target genomic reads
for pathogens or microbiomes <span class="citation" data-cites="10wcktDCE s3l2F2ef">[<a href="#ref-10wcktDCE" role="doc-biblioref">17</a>,<a href="#ref-s3l2F2ef" role="doc-biblioref">18</a>]</span>
has become particularly common, given its role in revealing widespread
infectious disease and possible epidemics that had previously been undetected in
the archaeological record <span class="citation" data-cites="1SX8gKlw XFBeyZB9 CNvvxsK7 1HEgHFzOs">[<a href="#ref-CNvvxsK7" role="doc-biblioref">11</a>,<a href="#ref-1HEgHFzOs" role="doc-biblioref">12</a>,<a href="#ref-1SX8gKlw" role="doc-biblioref">36</a>,<a href="#ref-XFBeyZB9" role="doc-biblioref">41</a>]</span>. Without easy access to the latest field-established
analytical routines, aDNA studies from groups new to the field risk being
published without the necessary quality control checks that ensure aDNA
authenticity and without yielding the full range of possibilities from their
data.</p>
<p>To address these shortcomings, we have completely re-implemented the latest
version of the EAGER pipeline in Nextflow <span class="citation" data-cites="4XDvZWxk">[<a href="#ref-4XDvZWxk" role="doc-biblioref">42</a>]</span> (a
domain-specific-language (DSL), specifically designed for the construction of
omics analysis pipelines), introduced new features, and more flexible pipeline
configurations. In addition, the newly named pipeline - nf-core/eager - has been
developed in the context of the nf-core community framework
<span class="citation" data-cites="wq4G2CfQ">[<a href="#ref-wq4G2CfQ" role="doc-biblioref">43</a>]</span>, which enforces strict guidelines for
best-practises in software development.</p>
<h1 id="results-and-discussion">Results and Discussion</h1>
<h2 id="scalability-portability-and-efficiency">Scalability, Portability, and Efficiency</h2>
<p>The reimplementation of EAGER into Nextflow offers a range of benefits over the
original custom pipeline framework.</p>
<p>Firstly, the new framework provides immediate integration of nf-core/eager into
various job schedulers in POSIX High-Performance-Cluster (HPC) environments,
cloud computing resources, and as well as local workstations. This portability
allows both small and big labs to run nf-core/eager regardless of the type of
computer or cluster size, with minimal effort or configuration, facilitating
reproducibility and therefore maintenance of standards within the field. This is
further assisted by the in-built compatibility with software environments and
containers such as Conda <span class="citation" data-cites="iU0FZYFP">[<a href="#ref-iU0FZYFP" role="doc-biblioref">44</a>]</span>,
Docker<span class="citation" data-cites="rmvJ2O0Q">[<a href="#ref-rmvJ2O0Q" role="doc-biblioref">45</a>]</span> and Singularity <span class="citation" data-cites="99QkJBoF">[<a href="#ref-99QkJBoF" role="doc-biblioref">46</a>]</span>.
These are single packages that include all the software (with exact versions)
required by the pipeline, in a form that is installable and runnable regardless
of the setup of their local software environment. Another major change with
nf-core/eager is that the graphical-user-interface (GUI) set up of an
nf-core/eager run is now replaced with a command-line-interface (CLI) as the
primary user interaction mode. This is more compatible and portable with most
HPCs (that may not offer display of program windows), and is in line with the
vast majority of bioinformatic tools. We therefore believe this will not be a
hindrance to new researchers from outside computational biology. However there
are plans from Nextflow (with tower.nf <span class="citation" data-cites="jJzDdoAE">[<a href="#ref-jJzDdoAE" role="doc-biblioref">47</a>]</span>) and within the
nf-core community to provide multiple alternatives in the near future including
a CLI wizard and a web-based input GUI.</p>
<p>Secondly, reproducibility is made easier through the use of ‘profiles’ that can
define configuration parameters. These profiles can be managed at different
hierarchical levels. HPC-level profiles can specify parameters for the computing
environment (job schedulers, cache locations for containers, maximum memory and
CPU resources etc.), which can be centrally managed to ensure all users of a
group use the same settings. Pipeline-level profiles, specifying default
parameters for nf-core/eager itself, allow fast access to routine pipeline-run
parameters via a single flag in the nf-core/eager run command, without having to
configure each new run from scratch.</p>
<p>Compared to the original EAGER that utilised per-FASTQ XML files with hardcoded
filepaths for a specific user’s server, nf-core/eager allows researchers to
publish the specific profile used in their runs alongside their publications, to
ensure other groups can generate the same results. Usage of profiles also
reduces mistakes caused by insufficient ‘prose’ based reporting of program
settings, which may occur in papers written by researchers unfamiliar with
informatics. The default nf-core/eager profile uses parameters evaluated in
different aDNA specific contexts (e.g. in <span class="citation" data-cites="WQbjOgBC"><a href="#ref-WQbjOgBC" role="doc-biblioref">48</a></span>], and
will be updated in each new release as new studies are published.</p>
<p>nf-core/eager provides improved efficiency over the original EAGER pipeline by
replacing the sample-by-sample sequential processing with Nextflow’s
asynchronous parallelisation, whereby multiple pipeline steps and samples are
run in parallel (in addition to single pipeline step multi-threading). This,
combined with pre-defined per-process customisation of resource parameters,
reduces unnecessary resource allocation that can occur with new users to each
step of an NGS data processing pipeline. This is particularly pertinent given
the increasing use of centralised HPCs or cloud computing, which often use
per-hour cost calculations.</p>
<h2 id="updated-workflow">Updated Workflow</h2>
<p>nf-core/eager follows a similar structural foundation to the original EAGER.
Given Illumina short-read FASTQ and/or BAM files, and a reference FASTA file,
this can be split into four main stages:</p>
<ol type="1">
<li>Preprocessing
<ul>
<li>Sequencing quality control: FastQC
<span class="citation" data-cites="R2o2aPyA">[<a href="#ref-R2o2aPyA" role="doc-biblioref">49</a>]</span></li>
<li>Sequencing artefact clean-up (merging, adapter clipping): AdapterRemoval2
<span class="citation" data-cites="L1GdB71G">[<a href="#ref-L1GdB71G" role="doc-biblioref">50</a>]</span></li>
<li>Preprocessing statistics generation</li>
</ul></li>
<li>Mapping and post- processing
<ul>
<li>Alignment against reference genome: BWA
<span class="citation" data-cites="roi5bwXL 1AzMUkwox">[<a href="#ref-roi5bwXL" role="doc-biblioref">39</a>,<a href="#ref-1AzMUkwox" role="doc-biblioref">51</a>]</span>, CircularMapper <span class="citation" data-cites="12TkhsLGD">[<a href="#ref-12TkhsLGD" role="doc-biblioref">38</a>]</span></li>
<li>Mapping quality filtering: Samtools <span class="citation" data-cites="IPc8cy7s">[<a href="#ref-IPc8cy7s" role="doc-biblioref">53</a>]</span></li>
<li>PCR duplicate removal: DeDup <span class="citation" data-cites="12TkhsLGD">[<a href="#ref-12TkhsLGD" role="doc-biblioref">38</a>]</span>, Picard
MarkDuplicates <span class="citation" data-cites="14nSa8zB9">[<a href="#ref-14nSa8zB9" role="doc-biblioref">54</a>]</span></li>
<li>Mapping statistics generation: PreSeq <span class="citation" data-cites="17zUqUTNA">[<a href="#ref-17zUqUTNA" role="doc-biblioref">55</a>]</span>, Qualimap2
<span class="citation" data-cites="O3E8cbBW">[<a href="#ref-O3E8cbBW" role="doc-biblioref">56</a>]</span></li>
</ul></li>
<li>aDNA Evaluation and Modification
<ul>
<li>Damage profiling: DamageProfiler <span class="citation" data-cites="1Ab7DZE7s">[<a href="#ref-1Ab7DZE7s" role="doc-biblioref">57</a>]</span></li>
<li>aDNA reads selection: PMDTools <span class="citation" data-cites="lh0Dclyl">[<a href="#ref-lh0Dclyl" role="doc-biblioref">58</a>]</span></li>
<li>Damage removal: Bamutils<span class="citation" data-cites="18bWeJ11I">[<a href="#ref-18bWeJ11I" role="doc-biblioref">59</a>]</span></li>
<li>(Human) contamination estimation: ANGSD <span class="citation" data-cites="FyiP2I2k">[<a href="#ref-FyiP2I2k" role="doc-biblioref">60</a>]</span></li>
</ul></li>
<li>Genotyping and Consensus Sequencing: GATK <span class="citation" data-cites="14nSa8zB9">[<a href="#ref-14nSa8zB9" role="doc-biblioref">54</a>]</span>,
VCF2Genome <span class="citation" data-cites="12TkhsLGD">[<a href="#ref-12TkhsLGD" role="doc-biblioref">38</a>]</span></li>
</ol>
<p>In nf-core/eager, all tools also originally used in EAGER have been updated to
latest versions, as available on Bioconda <span class="citation" data-cites="sYguBb3Q">[<a href="#ref-sYguBb3Q" role="doc-biblioref">61</a>]</span> and
Conda-forge <span class="citation" data-cites="qj0Rqpq8">[<a href="#ref-qj0Rqpq8" role="doc-biblioref">62</a>]</span> to ensure widespread
accessibility and stability of utilized tools. The MapDamage2 (for damage
profile generation) <span class="citation" data-cites="14YGEdSwc">[<a href="#ref-14YGEdSwc" role="doc-biblioref">35</a>]</span> and Schmutzi for
(mitochondrial contamination estimation) <span class="citation" data-cites="5S7TlWlb">[<a href="#ref-5S7TlWlb" role="doc-biblioref">63</a>]</span>
methods have not been carried over to nf-core/eager, the first because a more
performant successor method is now available (DamageProfiler), and the latter
because a stable release of the method could not be migrated to Bioconda. We
anticipate that there will be an updated version of Schmutzi in the near future
that will allow us to integrate the method again in nf-core/eager, once a
version is released on Bioconda. Support for the Bowtie2 aligner
<span class="citation" data-cites="PiS0h6Mu">[<a href="#ref-PiS0h6Mu" role="doc-biblioref">64</a>]</span> will be added in the near future, after consultation
with the palaeogenetics community. New tools to the basic workflow include fastp
<span class="citation" data-cites="pt9vlq4w">[<a href="#ref-pt9vlq4w" role="doc-biblioref">65</a>]</span> for the removal of ‘poly-G’ sequencing
artefacts that are common in 2-colour Illumina sequencing machines (such as the
increasingly popular NextSeq and NovaSeq platforms,
<span class="citation" data-cites="or1FajUp">[<a href="#ref-or1FajUp" role="doc-biblioref">66</a>]</span>).
For genotyping we have now included FreeBayes <span class="citation" data-cites="qKTLLip1">[<a href="#ref-qKTLLip1" role="doc-biblioref">67</a>]</span> as an
alternative to the human-focused GATK tools. We have also maintained the
possibility of using the now unsupported GATK UnifiedGenotyper, as the GATK
HaplotypeCaller performs <em>de novo</em> assembly around possible variants, which may
not be suitable for low-coverage aDNA data.</p>
<div id="fig:workflow-schematic" class="fignos">
<figure>
<img src="images/Fig1-eager2_workflow_v1.png" style="width:100.0%" alt="" /><figcaption><span>Figure 1:</span> Simplified schematic of the nf-core/eager workflow pipeline. Green filled
bubbles indicate new functionality added over the original EAGER
pipeline.</figcaption>
</figure>
</div>
<p>We have further extended the genomic analysis functionality of the pipeline by
adding ancient metagenomic analysis, to identify the wider taxonomic content of
a sample. We have added the ability to screen all off-target reads (not mapped
to the reference genome) with two metagenomic profilers: MALT
<span class="citation" data-cites="194vI7Tnf qDQBcgzB">[<a href="#ref-194vI7Tnf" role="doc-biblioref">68</a>,<a href="#ref-qDQBcgzB" role="doc-biblioref">69</a>]</span> and Kraken2
<span class="citation" data-cites="wfp4mBcT">[<a href="#ref-wfp4mBcT" role="doc-biblioref">70</a>]</span>. Characterisation of properties of authentic
aDNA from MALT alignments is carried out with the HOPS pipeline
<span class="citation" data-cites="cvMiJCKj">[<a href="#ref-cvMiJCKj" role="doc-biblioref">71</a>]</span>. Ancient metagenomic studies sometimes may
include comparative samples from living individuals
<span class="citation" data-cites="BF7nFrAi">[<a href="#ref-BF7nFrAi" role="doc-biblioref">72</a>]</span>. To support open data, whilst respecting data
privacy, nf-core/eager includes a ‘strip_fastq’ script which creates raw FASTQ
files, but with reference-genome mapped reads removed. This allows safe upload
of sequencing data to public repositories with identifiable human data removed.</p>
<p>Additional functionality tailored for ancient bacterial genomics includes
integration of a SNP alignment generation tool, MultiVCFAnalyzer
<span class="citation" data-cites="oasyTY16">[<a href="#ref-oasyTY16" role="doc-biblioref">8</a>]</span>, which allows assessment of cross-mapping levels from
different related taxa to a reference genome - a common challenge in ancient
bacterial genome reconstruction <span class="citation" data-cites="jz1XHWKM">[<a href="#ref-jz1XHWKM" role="doc-biblioref">34</a>]</span>. The
output SNP alignment FASTA file can then be used for downstream analyses such as
phylogenetic tree construction. Simple coverage statistics of particular
annotations (e.g. genes) of an input reference is offered by bedtools
<span class="citation" data-cites="1HWiAHnIw">[<a href="#ref-1HWiAHnIw" role="doc-biblioref">73</a>]</span>, which can be used, for example, for
determining functional differences between ancient bacterial strains (as in
<span class="citation" data-cites="XFBeyZB9">[<a href="#ref-XFBeyZB9" role="doc-biblioref">41</a>]</span>). When using a human reference genome,
nf-core/eager can now also can give estimates of the biological sex of a given
individual with Sex.DetEERRmine <span class="citation" data-cites="KlTB2lhf">[<a href="#ref-KlTB2lhf" role="doc-biblioref">74</a>]</span>. A dedicated
‘endogenous DNA’ calculator (endorS.py) is also included to provide a percentage
estimate of the sequenced reads matching the reference from the total number of
reads sequenced per library.</p>
<p>Given the large amount of sequencing often required to yield sufficient genome
coverage from aDNA data, palaeogeneticists tend to use multiple (differently
treated) libraries or sequencing runs. The original EAGER pipeline could only
run single libraries at a time, and in these contexts required significant
manual user input in merging different FASTQ or BAM files. A major upgrade in
nf-core/eager is that the new pipeline supports automated processing of complex
sequencing strategies for many samples. As an alternative to direct paths to
FASTQ or BAM files, the pipeline can also accepts a simple table in TSV format
which includes file paths and additional metadata such as sample name, library
name, sequencing lane, colour chemistry, and UDG treatment. This allows
simultaneous processing and appropriate merging of heterogeneous data from
multiple sequencing runs and/or library types.</p>
<p>The original EAGER pipeline required users to look through many independent
output directories and files to make full assessment of their sequencing data.
This has now been replaced with a much more extensive MultiQC
<span class="citation" data-cites="U9WbDW7I">[<a href="#ref-U9WbDW7I" role="doc-biblioref">75</a>]</span> report. This tool aggregates the log files
of every supported tool into a single interactive report, and assists users in
making fuller assessment of their sequencing and analysis runs. Most tools
within nf-core/eager have a corresponding MultiQC module to enable comprehensive
evaluation of all stages of the pipeline.</p>
<p>An overview of the entire pipeline is shown in Figure <a href="#fig:workflow-schematic">1</a>.</p>
<h2 id="accessibility">Accessibility</h2>
<p>Alongside the interactive MultiQC report, we have written extensive
documentation on all parts of running and interpreting the output of the
pipeline. Given that a large fraction of aDNA researchers come from fields
outside computational biology, and thus may have limited computational training,
we have written documentation that also gives guidance on how to interpret each
section of the report, specifically in the context of NGS and aDNA. This
includes schematic images of best practices or expected output that are
published under CC-BY licenses to allow for use in other training material
(example in <a href="#fig:doc-image-example">2</a>). We hope this open-access resource will
make the study of aDNA more accessible to researchers new to the field, by
providing practical guidelines how to evaluate characteristics and effects of
aDNA on downstream analyses.</p>
<div id="fig:doc-image-example" class="fignos">
<figure>
<img src="images/Fig2_fastqc_adapter_content.png" style="width:100.0%" alt="" /><figcaption><span>Figure 2:</span> Example of output interpretation documentation schematic images that can
assist new users in the interpretation to next-generation-sequencing aDNA
processing</figcaption>
</figure>
</div>
<p>The development of nf-core/eager in Nextflow and the nf-core initiative will
also improve open-source community contributions to the pipeline. While Nextflow
is written primarily in Groovy, the Nextflow DSL simplifies a number of concepts
to an intermediate level that bioinformaticians without Java/Groovy experience
can easily access (regardless of own programming language experience).
Furthermore, Nextflow places ubiquitous and more widely known command-line
interfaces, such as bash, in a prominent position within the code, rather than
custom java code and classes. We hope this will motivate further bug fixes and
feature contributions from the community, to keep the pipeline state-of-the-art
and ensure a longer life-cycle. This will also be supported by the active and
welcoming nf-core community who provide general guidance and advice on
developing Nextflow and nf-core pipelines.</p>
<h2 id="conclusion">Conclusion</h2>
<p>nf-core/eager is an efficient, portable, and accessible pipeline for processing
aDNA genomic data. This re-implementation of EAGER into Nextflow and nf-core
will improve reproducibility and inclusion of rapidly increasing aDNA datasets,
for both large and small laboratories. Extensive documentation also enables
newcomers to the field get a practical understanding on how to interpret aDNA in
the context of NGS data processing. Ultimately, nf-core/eager provides easier
access to the latest tools and routine screening analyses commonly used in the
field, and sets up the pipeline for staying at the forefront of palaeogenetic
analysis.</p>
<h2 id="methods">Methods</h2>
<h3 id="installation">Installation</h3>
<p>nf-core/eager requires a version of Java, Nextflow and either a functional Conda
installation <em>or</em> Docker/Singularity container installation. A quick
installation guide to follow to get started can be found in the <em>Quickstart</em>
section of the nf-core/eager repository
<span class="citation" data-cites="NXbAmDI0">[<a href="#ref-NXbAmDI0" role="doc-biblioref">76</a>]</span>.</p>
<h3 id="running">Running</h3>
<p>After the installation, users can run the pipeline using standard test data by
utilizing some of the <code>test</code> profiles we provide (e.g. using Docker):</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="ex">nextflow</span> run nf-core/eager -r 2.1.0 -profile test,docker</span></code></pre></div>
<p>This will download test data automatically, run the pipeline locally with all
software tools containerized in a Docker image and store the output of that run
in the <code>./results</code> folder of your current directory.</p>
<p>The default pipeline settings assumes paired end FASTQ data and will run:
FastQC; AdapterRemoval2 (merging and adapter clipping); post-clipping FastQC
(for AdapterRemoval2 performance evaluation); bwa mapping (with the ‘aln’
algorithm); samtools flagstat (for mapping statistics); endorS.py (for
endogenous DNA calculation); DeDup (for PCR amplicon deduplication); PreSeq (for
library complexity evaluation); DamageProfiler and Qualimap2 (for genome
coverage statistics) and the MultiQC pipeline run report. If no additional FASTA
indices are given, these will also be generated.</p>
<p>The pipeline is highly configurable and most modules can be turned on and off at
the request of the user using different flags to allow high customisation to
each users needs. For example, to include metagenomic screening of off-target
reads and sex determination based on on-target mappings of pre-clipped
single-end data:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="ex">nextflow</span> run nf-core/eager -r 2.1.0 -profile conda --input <span class="st">&#39;/&lt;path&gt;/&lt;to&gt;/*/*R1*.fastq.gz&#39;</span> --single_end --fasta <span class="st">&#39;/&lt;path&gt;/&lt;to&gt;/&lt;reference&gt;.fasta.gz&#39;</span> --skip_fastqc --skip_adapterremoval --run_bam_filtering --bam_discard_unmapped --bam_unmapped_type <span class="st">&#39;fastq&#39;</span> --run_metagenomic_screening --metagenomic_tool <span class="st">&#39;malt&#39;</span> --database <span class="st">&#39;/&lt;path&gt;/&lt;to&gt;/&lt;malt_database&gt;&#39;</span> --run_sexdeterrmine</span></code></pre></div>
<h4 id="profiles">Profiles</h4>
<p>We utilize a central configuration profile repository to enable users from
various institutions to use pipelines on their particular infrastructure more
easily <span class="citation" data-cites="13kS0x2y1">[<a href="#ref-13kS0x2y1" role="doc-biblioref">77</a>]</span>. There are multiple resources
listed in this repository with information on how to add your own configuration
profile with help from the nf-core community.</p>
<p>Users can customize this infrastructure profile by themselves, with the nf-core
community, or with their local system administrator to make sure that the
pipeline runs successfully, and can then rely on the Nextflow and nf-core
framework to ensure compatibility upon further infrastructure changes. For
example, in order to run the nf-core/eager pipeline at the Max Planck Institute
for the Science of Human History (MPI-SHH) in Jena, users only have to run:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="ex">nextflow</span> run nf-core/eager -r 2.1.0 -profile shh_cdag --input <span class="st">&#39;/&lt;path&gt;/&lt;to&gt;/*/*{R1,R2}*.fastq.gz&#39;</span> --fasta <span class="st">&#39;/&lt;path&gt;/&lt;to&gt;/&lt;reference&gt;.fasta.gz&#39;</span></span></code></pre></div>
<p>This runs the testing profile of the nf-core/eager pipeline with parameters
specifically adapted to the HPC system at the MPI-SHH. In some cases, similar
institutional configs for other institutions may already exist (originally
utilised for different nf-core pipelines), so users need not write their own.</p>
<h4 id="inputs">Inputs</h4>
<p>The pipeline can be started using (raw) FASTQ files from sequencing or
pre-mapped BAM files. Additionally, the pipeline requires a FASTA reference
genome.</p>
<p>If BAM input is provided, an optional conversion to FASTQ is offered, otherwise
BAM files processing will start from the post-mapping stage.</p>
<p>If users have complex set-ups, e.g. multiple sequencing lanes that require
merging of files for example, the pipeline can be supplied with a tabular
separated value (TSV) file to enable such complex data handling. Both FASTQs and
BAMs can be provided in this set up. FASTQs with the same library name and
sequencing chemistry but sequenced across multiple lanes will be concatenated
after AdapterRemoval and prior mapping. Libraries with the sample name and with
the same UDG treatment, will be merged after deduplication. If libraries with
the sample name have different UDG treatment, these will be merged after the
aDNA modification stage (i.e. BAM trimming or PMDtools, if turned on), prior
genotyping.</p>
<p>As Nextflow will automatically download files from URLs, profiles and/or TSV
files can include links to publicly available data (e.g. the ENA FTP server).
This assists in reproducibility as if profiles or TSV files are uploaded with a
publication, a researcher wishing to re-analyse the data in the same way can use
the exact settings and merging procedures in the original publication,
without having to reconstruct this from prose.</p>
<h4 id="monitoring">Monitoring</h4>
<p>Users can either monitor their pipeline execution with the messages Nextflow
prints to the console while running, or utilize projects such as Nextflow Tower
<span class="citation" data-cites="jJzDdoAE">[<a href="#ref-jJzDdoAE" role="doc-biblioref">47</a>]</span> to monitor their analysis pipeline during runtime.</p>
<h4 id="output">Output</h4>
<p>The pipeline produces several dozen output files in various file formats, with a
more detailed listing available in the user documentation. This includes
metrics, statistical analysis data and standardized output files (BAM, VCF) for
close inspection and further downstream analysis, as well as a MultiQC report.
If an emailing daemon is set up on the server, the latter can even be emailed to
users automatically, when starting the pipeline with a dedicated option (–email
you@yourdomain.org).</p>
<h2 id="data-and-software-availability">Data and software availability</h2>
<p>All code is available on github at
<a href="https://github.com/nf-core/eager">https://github.com/nf-core/eager</a> and
archived with Zenodo under the DOI
<a href="https://doi.org/10.5281/zenodo.1465061">10.5281/zenodo.1465061</a>. All test data
is from the ENA public repository available under ENA IDs: <strong>FIXME</strong></p>
<p>This paper was written with Manubot <span class="citation" data-cites="YuJbg3zO">[<a href="#ref-YuJbg3zO" role="doc-biblioref">78</a>]</span>.</p>
<h2 id="competing-interests">Competing Interests</h2>
<p>No competing interests are declared.</p>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>We thank the nf-core community for general support and suggestions during the
writing of the pipeline. We also thank Arielle Munters, Hester van Schalkwyk,
Irina Velsko, Katerine Eaton, Luc Venturini, Marcel Keller, Pierre Lindenbaum,
Pontus Skoglund, Raphael Eisenhofer, Torsten Günter, and Kevin Lord for bug
reports and feature suggestions. We are grateful to the members of the
Department of Archaeogenetics at the Max Planck Institute for the Science of
Human History who performed beta testing of the pipeline. We thank the aDNA
twitter community for responding to polls regarding design decisions of the
pipeline.</p>
<p>We also want to thank Alexander Herbig and Wolfgang Haak for kindly providing
comments on this manuscript.</p>
<p>We also want to acknowledge Christina Warinner, Stephan Schiffels and the Max Planck
Society who provided funds for travel to nf-core events.</p>
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references hanging-indent" role="doc-bibliography">
<div id="ref-86EfjHO2">
<p>1. <strong>Complete Genomes Reveal Signatures of Demographic and Genetic Declines in the Woolly Mammoth</strong><br />
Eleftheria Palkopoulou, Swapan Mallick, Pontus Skoglund, Jacob Enk, Nadin Rohland, Heng Li, Ayça Omrak, Sergey Vartanyan, Hendrik Poinar, Anders Götherström, … Love Dalén<br />
<em>Current Biology</em> (2015-05) <a href="https://doi.org/34d">https://doi.org/34d</a><br />
DOI: <a href="https://doi.org/10.1016/j.cub.2015.04.007">10.1016/j.cub.2015.04.007</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25913407">25913407</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4439331">PMC4439331</a></p>
</div>
<div id="ref-1EJZxEzq1">
<p>2. <strong>Recalibrating Equus evolution using the genome sequence of an early Middle Pleistocene horse</strong><br />
Ludovic Orlando, Aurélien Ginolhac, Guojie Zhang, Duane Froese, Anders Albrechtsen, Mathias Stiller, Mikkel Schubert, Enrico Cappellini, Bent Petersen, Ida Moltke, … Eske Willerslev<br />
<em>Nature</em> (2013-06-26) <a href="https://doi.org/q7n">https://doi.org/q7n</a><br />
DOI: <a href="https://doi.org/10.1038/nature12323">10.1038/nature12323</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23803765">23803765</a></p>
</div>
<div id="ref-7nLP30p0">
<p>3. <strong>Ancient pigs reveal a near-complete genomic turnover following their introduction to Europe</strong><br />
Laurent A. F. Frantz, James Haile, Audrey T. Lin, Amelie Scheu, Christina Geörg, Norbert Benecke, Michelle Alexander, Anna Linderholm, Victoria E. Mullin, Kevin G. Daly, … Greger Larson<br />
<em>Proceedings of the National Academy of Sciences</em> (2019-08-27) <a href="https://doi.org/gf9hnf">https://doi.org/gf9hnf</a><br />
DOI: <a href="https://doi.org/10.1073/pnas.1901169116">10.1073/pnas.1901169116</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31405970">31405970</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6717267">PMC6717267</a></p>
</div>
<div id="ref-ZMf8LvSn">
<p>4. <strong>Ancient DNA reveals the Arctic origin of Viking Age cod from Haithabu, Germany</strong><br />
Bastiaan Star, Sanne Boessenkool, Agata T. Gondek, Elena A. Nikulina, Anne Karin Hufthammer, Christophe Pampoulie, Halvor Knutsen, Carl André, Heidi M. Nistelberger, Jan Dierking, … James H. Barrett<br />
<em>Proceedings of the National Academy of Sciences</em> (2017-08-22) <a href="https://doi.org/gbt8b2">https://doi.org/gbt8b2</a><br />
DOI: <a href="https://doi.org/10.1073/pnas.1710186114">10.1073/pnas.1710186114</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28784790">28784790</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5576834">PMC5576834</a></p>
</div>
<div id="ref-tAnEwrSn">
<p>5. <strong>137 ancient human genomes from across the Eurasian steppes</strong><br />
Peter de Barros Damgaard, Nina Marchi, Simon Rasmussen, Michaël Peyrot, Gabriel Renaud, Thorfinn Korneliussen, J. Víctor Moreno-Mayar, Mikkel Winther Pedersen, Amy Goldberg, Emma Usmanova, … Eske Willerslev<br />
<em>Nature</em> (2018-05-09) <a href="https://doi.org/gd8hs5">https://doi.org/gd8hs5</a><br />
DOI: <a href="https://doi.org/10.1038/s41586-018-0094-2">10.1038/s41586-018-0094-2</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29743675">29743675</a></p>
</div>
<div id="ref-dHnkKsKe">
<p>6. <strong>A High-Coverage Genome Sequence from an Archaic Denisovan Individual</strong><br />
M. Meyer, M. Kircher, M.-T. Gansauge, H. Li, F. Racimo, S. Mallick, J. G. Schraiber, F. Jay, K. Prufer, C. de Filippo, … S. Paabo<br />
<em>Science</em> (2012-08-30) <a href="https://doi.org/q8p">https://doi.org/q8p</a><br />
DOI: <a href="https://doi.org/10.1126/science.1224344">10.1126/science.1224344</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22936568">22936568</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3617501">PMC3617501</a></p>
</div>
<div id="ref-qngWumct">
<p>7. <strong>The genome of the offspring of a Neanderthal mother and a Denisovan father</strong><br />
Viviane Slon, Fabrizio Mafessoni, Benjamin Vernot, Cesare de Filippo, Steffi Grote, Bence Viola, Mateja Hajdinjak, Stéphane Peyrégne, Sarah Nagel, Samantha Brown, … Svante Pääbo<br />
<em>Nature</em> (2018-08-22) <a href="https://doi.org/cs64">https://doi.org/cs64</a><br />
DOI: <a href="https://doi.org/10.1038/s41586-018-0455-x">10.1038/s41586-018-0455-x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30135579">30135579</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6130845">PMC6130845</a></p>
</div>
<div id="ref-oasyTY16">
<p>8. <strong>Pre-Columbian mycobacterial genomes reveal seals as a source of New World human tuberculosis</strong><br />
Kirsten I. Bos, Kelly M. Harkins, Alexander Herbig, Mireia Coscolla, Nico Weber, Iñaki Comas, Stephen A. Forrest, Josephine M. Bryant, Simon R. Harris, Verena J. Schuenemann, … Johannes Krause<br />
<em>Nature</em> (2014-08-20) <a href="https://doi.org/f6nk4g">https://doi.org/f6nk4g</a><br />
DOI: <a href="https://doi.org/10.1038/nature13591">10.1038/nature13591</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25141181">25141181</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4550673">PMC4550673</a></p>
</div>
<div id="ref-1APjJqbjj">
<p>9. <strong>Integrative approach using <em>Yersinia pestis</em> genomes to revisit the historical landscape of plague during the Medieval Period</strong><br />
Amine Namouchi, Meriam Guellil, Oliver Kersten, Stephanie Hänsch, Claudio Ottoni, Boris V. Schmid, Elsa Pacciani, Luisa Quaglia, Marco Vermunt, Egil L. Bauer, … Barbara Bramanti<br />
<em>Proceedings of the National Academy of Sciences</em> (2018-12-11) <a href="https://doi.org/ggfn3h">https://doi.org/ggfn3h</a><br />
DOI: <a href="https://doi.org/10.1073/pnas.1812865115">10.1073/pnas.1812865115</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30478041">30478041</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6294933">PMC6294933</a></p>
</div>
<div id="ref-4YuC8sCQ">
<p>10. <strong>Ancient genomes reveal a high diversity of Mycobacterium leprae in medieval Europe</strong><br />
Verena J. Schuenemann, Charlotte Avanzi, Ben Krause-Kyora, Alexander Seitz, Alexander Herbig, Sarah Inskip, Marion Bonazzi, Ella Reiter, Christian Urban, Dorthe Dangvard Pedersen, … Johannes Krause<br />
<em>PLOS Pathogens</em> (2018-05-10) <a href="https://doi.org/gdrj4v">https://doi.org/gdrj4v</a><br />
DOI: <a href="https://doi.org/10.1371/journal.ppat.1006997">10.1371/journal.ppat.1006997</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29746563">29746563</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5944922">PMC5944922</a></p>
</div>
<div id="ref-CNvvxsK7">
<p>11. <strong>Ancient hepatitis B viruses from the Bronze Age to the Medieval period</strong><br />
Barbara Mühlemann, Terry C. Jones, Peter de Barros Damgaard, Morten E. Allentoft, Irina Shevnina, Andrey Logvin, Emma Usmanova, Irina P. Panyushkina, Bazartseren Boldgiv, Tsevel Bazartseren, … Eske Willerslev<br />
<em>Nature</em> (2018-05-09) <a href="https://doi.org/gddxvj">https://doi.org/gddxvj</a><br />
DOI: <a href="https://doi.org/10.1038/s41586-018-0097-z">10.1038/s41586-018-0097-z</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29743673">29743673</a></p>
</div>
<div id="ref-1HEgHFzOs">
<p>12. <strong>Neolithic and medieval virus genomes reveal complex evolution of hepatitis B</strong><br />
Ben Krause-Kyora, Julian Susat, Felix M Key, Denise Kühnert, Esther Bosse, Alexander Immel, Christoph Rinne, Sabin-Christin Kornell, Diego Yepes, Sören Franzenburg, … Johannes Krause<br />
<em>eLife</em> (2018-05-10) <a href="https://doi.org/gdhck2">https://doi.org/gdhck2</a><br />
DOI: <a href="https://doi.org/10.7554/elife.36666">10.7554/elife.36666</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29745896">29745896</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6008052">PMC6008052</a></p>
</div>
<div id="ref-1001q664T">
<p>13. <strong>Ancient reveals the timing and persistence of organellar genetic bottlenecks over 3,000 years of sunflower domestication and improvement</strong><br />
Nathan Wales, Melis Akman, Ray H. B. Watson, Fátima Sánchez Barreiro, Bruce D. Smith, Kristen J. Gremillion, M. Thomas P. Gilbert, Benjamin K. Blackman<br />
<em>Evolutionary Applications</em> (2018-02-13) <a href="https://doi.org/gf568v">https://doi.org/gf568v</a><br />
DOI: <a href="https://doi.org/10.1111/eva.12594">10.1111/eva.12594</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30622634">30622634</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6304678">PMC6304678</a></p>
</div>
<div id="ref-6MKvilVe">
<p>14. <strong>The origins and adaptation of European potatoes reconstructed from historical genomes</strong><br />
Rafal M. Gutaker, Clemens L. Weiß, David Ellis, Noelle L. Anglin, Sandra Knapp, José Luis Fernández-Alonso, Salomé Prat, Hernán A. Burbano<br />
<em>Nature Ecology &amp; Evolution</em> (2019-06-24) <a href="https://doi.org/ggxkk8">https://doi.org/ggxkk8</a><br />
DOI: <a href="https://doi.org/10.1038/s41559-019-0921-3">10.1038/s41559-019-0921-3</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31235927">31235927</a></p>
</div>
<div id="ref-fK2md3DI">
<p>15. <strong>The Prevotella copri Complex Comprises Four Distinct Clades Underrepresented in Westernized Populations</strong><br />
Adrian Tett, Kun D. Huang, Francesco Asnicar, Hannah Fehlner-Peach, Edoardo Pasolli, Nicolai Karcher, Federica Armanini, Paolo Manghi, Kevin Bonham, Moreno Zolfo, … Nicola Segata<br />
<em>Cell Host &amp; Microbe</em> (2019-11) <a href="https://doi.org/ggc9dc">https://doi.org/ggc9dc</a><br />
DOI: <a href="https://doi.org/10.1016/j.chom.2019.08.018">10.1016/j.chom.2019.08.018</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31607556">31607556</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6854460">PMC6854460</a></p>
</div>
<div id="ref-h6xmK5cY">
<p>16. <strong>CoproID predicts the source of coprolites and paleofeces using microbiome composition and host DNA content</strong><br />
Maxime Borry, Bryan Cordova, Angela Perri, Marsha Wibowo, Tanvi Prasad Honap, Jada Ko, Jie Yu, Kate Britton, Linus Girdland-Flink, Robert C. Power, … Christina Warinner<br />
<em>PeerJ</em> (2020-04-17) <a href="https://doi.org/dr8x">https://doi.org/dr8x</a><br />
DOI: <a href="https://doi.org/10.7717/peerj.9001">10.7717/peerj.9001</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32337106">32337106</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7169968">PMC7169968</a></p>
</div>
<div id="ref-10wcktDCE">
<p>17. <strong>Pathogens and host immunity in the ancient human oral cavity</strong><br />
Christina Warinner, João F Matias Rodrigues, Rounak Vyas, Christian Trachsel, Natallia Shved, Jonas Grossmann, Anita Radini, Y Hancock, Raul Y Tito, Sarah Fiddyment, … Enrico Cappellini<br />
<em>Nature Genetics</em> (2014-02-23) <a href="https://doi.org/r4n">https://doi.org/r4n</a><br />
DOI: <a href="https://doi.org/10.1038/ng.2906">10.1038/ng.2906</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24562188">24562188</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3969750">PMC3969750</a></p>
</div>
<div id="ref-s3l2F2ef">
<p>18. <strong>Neanderthal behaviour, diet, and disease inferred from ancient DNA in dental calculus</strong><br />
Laura S. Weyrich, Sebastian Duchene, Julien Soubrier, Luis Arriola, Bastien Llamas, James Breen, Alan G. Morris, Kurt W. Alt, David Caramelli, Veit Dresely, … Alan Cooper<br />
<em>Nature</em> (2017-03-08) <a href="https://doi.org/f9szrm">https://doi.org/f9szrm</a><br />
DOI: <a href="https://doi.org/10.1038/nature21674">10.1038/nature21674</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28273061">28273061</a></p>
</div>
<div id="ref-YBHwLDej">
<p>19. <strong>Fifty thousand years of Arctic vegetation and megafaunal diet</strong><br />
Eske Willerslev, John Davison, Mari Moora, Martin Zobel, Eric Coissac, Mary E. Edwards, Eline D. Lorenzen, Mette Vestergård, Galina Gussarova, James Haile, … Pierre Taberlet<br />
<em>Nature</em> (2014-02-05) <a href="https://doi.org/f2zr4s">https://doi.org/f2zr4s</a><br />
DOI: <a href="https://doi.org/10.1038/nature12921">10.1038/nature12921</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24499916">24499916</a></p>
</div>
<div id="ref-64ZHuA4N">
<p>20. <strong>Neandertal and Denisovan DNA from Pleistocene sediments</strong><br />
Viviane Slon, Charlotte Hopfe, Clemens L. Weiß, Fabrizio Mafessoni, Marco de la Rasilla, Carles Lalueza-Fox, Antonio Rosas, Marie Soressi, Monika V. Knul, Rebecca Miller, … Matthias Meyer<br />
<em>Science</em> (2017-05-12) <a href="https://doi.org/b6jd">https://doi.org/b6jd</a><br />
DOI: <a href="https://doi.org/10.1126/science.aam9695">10.1126/science.aam9695</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28450384">28450384</a></p>
</div>
<div id="ref-WIkaempF">
<p>21. <strong>Plasmodium vivax Malaria Viewed through the Lens of an Eradicated European Strain</strong><br />
Lucy van Dorp, Pere Gelabert, Adrien Rieux, Marc de Manuel, Toni de-Dios, Shyam Gopalakrishnan, Christian Carøe, Marcela Sandoval-Velasco, Rosa Fregel, Iñigo Olalde, … Carles Lalueza-Fox<br />
<em>Molecular Biology and Evolution</em> (2020-03) <a href="https://doi.org/ggqzq2">https://doi.org/ggqzq2</a><br />
DOI: <a href="https://doi.org/10.1093/molbev/msz264">10.1093/molbev/msz264</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31697387">31697387</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7038659">PMC7038659</a></p>
</div>
<div id="ref-M0SKmPyu">
<p>22. <strong>Paging through history: parchment as a reservoir of ancient DNA for next generation sequencing</strong><br />
M. D. Teasdale, N. L. van Doorn, S. Fiddyment, C. C. Webb, T. O’Connor, M. Hofreiter, M. J. Collins, D. G. Bradley<br />
<em>Philosophical Transactions of the Royal Society B: Biological Sciences</em> (2015-01-19) <a href="https://doi.org/ggqzq3">https://doi.org/ggqzq3</a><br />
DOI: <a href="https://doi.org/10.1098/rstb.2013.0379">10.1098/rstb.2013.0379</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25487331">25487331</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4275887">PMC4275887</a></p>
</div>
<div id="ref-NXcxgeob">
<p>23. <strong>A 5700 year-old human genome and oral microbiome from chewed birch pitch</strong><br />
Theis Z. T. Jensen, Jonas Niemann, Katrine Højholt Iversen, Anna K. Fotakis, Shyam Gopalakrishnan, Åshild J. Vågene, Mikkel Winther Pedersen, Mikkel-Holger S. Sinding, Martin R. Ellegaard, Morten E. Allentoft, … Hannes Schroeder<br />
<em>Nature Communications</em> (2019-12-17) <a href="https://doi.org/ggfm6x">https://doi.org/ggfm6x</a><br />
DOI: <a href="https://doi.org/10.1038/s41467-019-13549-9">10.1038/s41467-019-13549-9</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31848342">31848342</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6917805">PMC6917805</a></p>
</div>
<div id="ref-10155v3VM">
<p>24. <strong>Ancient DNA from mastics solidifies connection between material culture and genetics of mesolithic hunter–gatherers in Scandinavia</strong><br />
Natalija Kashuba, Emrah Kırdök, Hege Damlien, Mikael A. Manninen, Bengt Nordqvist, Per Persson, Anders Götherström<br />
<em>Communications Biology</em> (2019-05-15) <a href="https://doi.org/ggqzqz">https://doi.org/ggqzqz</a><br />
DOI: <a href="https://doi.org/10.1038/s42003-019-0399-1">10.1038/s42003-019-0399-1</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31123709">31123709</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6520363">PMC6520363</a></p>
</div>
<div id="ref-aSy3Srju">
<p>25. <strong>The Beaker phenomenon and the genomic transformation of northwest Europe</strong><br />
Iñigo Olalde, Selina Brace, Morten E. Allentoft, Ian Armit, Kristian Kristiansen, Thomas Booth, Nadin Rohland, Swapan Mallick, Anna Szécsényi-Nagy, Alissa Mittnik, … David Reich<br />
<em>Nature</em> (2018-02-21) <a href="https://doi.org/gcx74m">https://doi.org/gcx74m</a><br />
DOI: <a href="https://doi.org/10.1038/nature25738">10.1038/nature25738</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29466337">29466337</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5973796">PMC5973796</a></p>
</div>
<div id="ref-wT5SMjro">
<p>26. <strong>The genomic history of southeastern Europe</strong><br />
Iain Mathieson, Songül Alpaslan-Roodenberg, Cosimo Posth, Anna Szécsényi-Nagy, Nadin Rohland, Swapan Mallick, Iñigo Olalde, Nasreen Broomandkhoshbacht, Francesca Candilio, Olivia Cheronet, … David Reich<br />
<em>Nature</em> (2018-02-21) <a href="https://doi.org/gc2n9h">https://doi.org/gc2n9h</a><br />
DOI: <a href="https://doi.org/10.1038/nature25778">10.1038/nature25778</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29466330">29466330</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6091220">PMC6091220</a></p>
</div>
<div id="ref-1AkUQBPuG">
<p>27. <strong>A draft genome of Yersinia pestis from victims of the Black Death</strong><br />
Kirsten I. Bos, Verena J. Schuenemann, G. Brian Golding, Hernán A. Burbano, Nicholas Waglechner, Brian K. Coombes, Joseph B. McPhee, Sharon N. DeWitte, Matthias Meyer, Sarah Schmedes, … Johannes Krause<br />
<em>Nature</em> (2011-10-12) <a href="https://doi.org/fk87wk">https://doi.org/fk87wk</a><br />
DOI: <a href="https://doi.org/10.1038/nature10549">10.1038/nature10549</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/21993626">21993626</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3690193">PMC3690193</a></p>
</div>
<div id="ref-TwlGYuW7">
<p>28. <strong>Bioinformatics Education–Perspectives and Challenges out of Africa</strong><br />
O. Tastan Bishop, E. F. Adebiyi, A. M. Alzohairy, D. Everett, K. Ghedira, A. Ghouila, J. Kumuthini, N. J. Mulder, S. Panji, H.-G. Patterton, (for the H3ABioNet Consortium, as members of The H3Africa Consortium)<br />
<em>Briefings in Bioinformatics</em> (2014-07-02) <a href="https://doi.org/f67hjx">https://doi.org/f67hjx</a><br />
DOI: <a href="https://doi.org/10.1093/bib/bbu022">10.1093/bib/bbu022</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24990350">24990350</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4364068">PMC4364068</a></p>
</div>
<div id="ref-h1qNDSRd">
<p>29. <strong>Highlights on the Application of Genomics and Bioinformatics in the Fight Against Infectious Diseases: Challenges and Opportunities in Africa</strong><br />
Saikou Y. Bah, Collins Misita Morang’a, Jonas A. Kengne-Ouafo, Lucas Amenga–Etego, Gordon A. Awandare<br />
<em>Frontiers in Genetics</em> (2018-11-27) <a href="https://doi.org/gfrxbz">https://doi.org/gfrxbz</a><br />
DOI: <a href="https://doi.org/10.3389/fgene.2018.00575">10.3389/fgene.2018.00575</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30538723">30538723</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6277583">PMC6277583</a></p>
</div>
<div id="ref-Zp5bQCLJ">
<p>30. <strong>Instability and decay of the primary structure of DNA</strong><br />
Tomas Lindahl<br />
<em>Nature</em> (1993-04) <a href="https://doi.org/d9c9vq">https://doi.org/d9c9vq</a><br />
DOI: <a href="https://doi.org/10.1038/362709a0">10.1038/362709a0</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/8469282">8469282</a></p>
</div>
<div id="ref-ZeQp0m3j">
<p>31. <strong>Nuclear DNA sequences from the Middle Pleistocene Sima de los Huesos hominins</strong><br />
Matthias Meyer, Juan-Luis Arsuaga, Cesare de Filippo, Sarah Nagel, Ayinuer Aximu-Petri, Birgit Nickel, Ignacio Martínez, Ana Gracia, José María Bermúdez de Castro, Eudald Carbonell, … Svante Pääbo<br />
<em>Nature</em> (2016-03-14) <a href="https://doi.org/bdcn">https://doi.org/bdcn</a><br />
DOI: <a href="https://doi.org/10.1038/nature17405">10.1038/nature17405</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26976447">26976447</a></p>
</div>
<div id="ref-dbfFgAgr">
<p>32. <strong>Patterns of damage in genomic DNA sequences from a Neandertal</strong><br />
A. W. Briggs, U. Stenzel, P. L. F. Johnson, R. E. Green, J. Kelso, K. Prufer, M. Meyer, J. Krause, M. T. Ronan, M. Lachmann, S. Paabo<br />
<em>Proceedings of the National Academy of Sciences</em> (2007-08-21) <a href="https://doi.org/bs4w7h">https://doi.org/bs4w7h</a><br />
DOI: <a href="https://doi.org/10.1073/pnas.0704665104">10.1073/pnas.0704665104</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/17715061">17715061</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1976210">PMC1976210</a></p>
</div>
<div id="ref-NVqMN9QU">
<p>33. <strong>A new model for ancient DNA decay based on paleogenomic meta-analysis</strong><br />
Logan Kistler, Roselyn Ware, Oliver Smith, Matthew Collins, Robin G. Allaby<br />
<em>Nucleic Acids Research</em> (2017-06-20) <a href="https://doi.org/gf58ts">https://doi.org/gf58ts</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkx361">10.1093/nar/gkx361</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28486705">28486705</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5499742">PMC5499742</a></p>
</div>
<div id="ref-jz1XHWKM">
<p>34. <strong>A Robust Framework for Microbial Archaeology</strong><br />
Christina Warinner, Alexander Herbig, Allison Mann, James A. Fellows Yates, Clemens L. Weiß, Hernán A. Burbano, Ludovic Orlando, Johannes Krause<br />
<em>Annual Review of Genomics and Human Genetics</em> (2017-08-31) <a href="https://doi.org/gf5wqv">https://doi.org/gf5wqv</a><br />
DOI: <a href="https://doi.org/10.1146/annurev-genom-091416-035526">10.1146/annurev-genom-091416-035526</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28460196">28460196</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5581243">PMC5581243</a></p>
</div>
<div id="ref-14YGEdSwc">
<p>35. <strong>mapDamage2.0: fast approximate Bayesian estimates of ancient DNA damage parameters</strong><br />
Hákon Jónsson, Aurélien Ginolhac, Mikkel Schubert, Philip L. F. Johnson, Ludovic Orlando<br />
<em>Bioinformatics</em> (2013-07) <a href="https://doi.org/gb5g2t">https://doi.org/gb5g2t</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btt193">10.1093/bioinformatics/btt193</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23613487">23613487</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3694634">PMC3694634</a></p>
</div>
<div id="ref-1SX8gKlw">
<p>36. <strong>Early Divergent Strains of Yersinia pestis in Eurasia 5,000 Years Ago</strong><br />
Simon Rasmussen, Morten Erik Allentoft, Kasper Nielsen, Ludovic Orlando, Martin Sikora, Karl-Göran Sjögren, Anders Gorm Pedersen, Mikkel Schubert, Alex Van Dam, Christian Moliin Outzen Kapel, … Eske Willerslev<br />
<em>Cell</em> (2015-10) <a href="https://doi.org/f3mxqd">https://doi.org/f3mxqd</a><br />
DOI: <a href="https://doi.org/10.1016/j.cell.2015.10.009">10.1016/j.cell.2015.10.009</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26496604">26496604</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4644222">PMC4644222</a></p>
</div>
<div id="ref-1H3banOxK">
<p>37. <strong>Characterization of ancient and modern genomes by SNP detection and phylogenomic and metagenomic analysis using PALEOMIX</strong><br />
Mikkel Schubert, Luca Ermini, Clio Der Sarkissian, Hákon Jónsson, Aurélien Ginolhac, Robert Schaefer, Michael D Martin, Ruth Fernández, Martin Kircher, Molly McCue, … Ludovic Orlando<br />
<em>Nature Protocols</em> (2014-04-10) <a href="https://doi.org/f5x3qm">https://doi.org/f5x3qm</a><br />
DOI: <a href="https://doi.org/10.1038/nprot.2014.063">10.1038/nprot.2014.063</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24722405">24722405</a></p>
</div>
<div id="ref-12TkhsLGD">
<p>38. <strong>EAGER: efficient ancient genome reconstruction</strong><br />
Alexander Peltzer, Günter Jäger, Alexander Herbig, Alexander Seitz, Christian Kniep, Johannes Krause, Kay Nieselt<br />
<em>Genome Biology</em> (2016-03-31) <a href="https://doi.org/ggqzpk">https://doi.org/ggqzpk</a><br />
DOI: <a href="https://doi.org/10.1186/s13059-016-0918-z">10.1186/s13059-016-0918-z</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27036623">27036623</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4815194">PMC4815194</a></p>
</div>
<div id="ref-roi5bwXL">
<p>39. <strong>Fast and accurate short read alignment with Burrows-Wheeler transform</strong><br />
H. Li, R. Durbin<br />
<em>Bioinformatics</em> (2009-05-18) <a href="https://doi.org/dqt59j">https://doi.org/dqt59j</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btp324">10.1093/bioinformatics/btp324</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19451168">19451168</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2705234">PMC2705234</a></p>
</div>
<div id="ref-Rhiz5BNL">
<p>40. <strong>mapDamage: testing for damage patterns in ancient DNA sequences</strong><br />
Aurelien Ginolhac, Morten Rasmussen, M. Thomas P. Gilbert, Eske Willerslev, Ludovic Orlando<br />
<em>Bioinformatics</em> (2011-08-01) <a href="https://doi.org/cn45v7">https://doi.org/cn45v7</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btr347">10.1093/bioinformatics/btr347</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/21659319">21659319</a></p>
</div>
<div id="ref-XFBeyZB9">
<p>41. <strong>The Stone Age Plague and Its Persistence in Eurasia</strong><br />
Aida Andrades Valtueña, Alissa Mittnik, Felix M. Key, Wolfgang Haak, Raili Allmäe, Andrej Belinskij, Mantas Daubaras, Michal Feldman, Rimantas Jankauskas, Ivor Janković, … Johannes Krause<br />
<em>Current Biology</em> (2017-12) <a href="https://doi.org/cgmv">https://doi.org/cgmv</a><br />
DOI: <a href="https://doi.org/10.1016/j.cub.2017.10.025">10.1016/j.cub.2017.10.025</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29174893">29174893</a></p>
</div>
<div id="ref-4XDvZWxk">
<p>42. <strong>Nextflow enables reproducible computational workflows</strong><br />
Paolo Di Tommaso, Maria Chatzou, Evan W Floden, Pablo Prieto Barja, Emilio Palumbo, Cedric Notredame<br />
<em>Nature Biotechnology</em> (2017-04-11) <a href="https://doi.org/gfj52z">https://doi.org/gfj52z</a><br />
DOI: <a href="https://doi.org/10.1038/nbt.3820">10.1038/nbt.3820</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28398311">28398311</a></p>
</div>
<div id="ref-wq4G2CfQ">
<p>43. <strong>The nf-core framework for community-curated bioinformatics pipelines</strong><br />
Philip A. Ewels, Alexander Peltzer, Sven Fillinger, Harshil Patel, Johannes Alneberg, Andreas Wilm, Maxime Ulysse Garcia, Paolo Di Tommaso, Sven Nahnsen<br />
<em>Nature Biotechnology</em> (2020-02-13) <a href="https://doi.org/ggk3qh">https://doi.org/ggk3qh</a><br />
DOI: <a href="https://doi.org/10.1038/s41587-020-0439-x">10.1038/s41587-020-0439-x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32055031">32055031</a></p>
</div>
<div id="ref-iU0FZYFP">
<p>44. <strong>Conda — Conda documentation</strong><a href="https://docs.conda.io/en/latest/">https://docs.conda.io/en/latest/</a></p>
</div>
<div id="ref-rmvJ2O0Q">
<p>45. <strong>Empowering App Development for Developers | Docker</strong><a href="https://www.docker.com/">https://www.docker.com/</a></p>
</div>
<div id="ref-99QkJBoF">
<p>46. <strong>Home</strong><br />
Sylabs.io<br />
<a href="https://sylabs.io/">https://sylabs.io/</a></p>
</div>
<div id="ref-jJzDdoAE">
<p>47. <strong>Nextflow Tower</strong><a href="https://tower.nf/">https://tower.nf/</a></p>
</div>
<div id="ref-WQbjOgBC">
<p>48. <strong>Improving ancient DNA read mapping against modern reference genomes</strong><br />
Mikkel Schubert, Aurelien Ginolhac, Stinus Lindgreen, John F Thompson, Khaled AS AL-Rasheid, Eske Willerslev, Anders Krogh, Ludovic Orlando<br />
<em>BMC Genomics</em> (2012) <a href="https://doi.org/gb3ff7">https://doi.org/gb3ff7</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2164-13-178">10.1186/1471-2164-13-178</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22574660">22574660</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3468387">PMC3468387</a></p>
</div>
<div id="ref-R2o2aPyA">
<p>49. <strong>Babraham Bioinformatics - FastQC A Quality Control tool for High Throughput Sequence Data</strong><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></p>
</div>
<div id="ref-L1GdB71G">
<p>50. <strong>AdapterRemoval v2: rapid adapter trimming, identification, and read merging</strong><br />
Mikkel Schubert, Stinus Lindgreen, Ludovic Orlando<br />
<em>BMC Research Notes</em> (2016-02-12) <a href="https://doi.org/gfzqhb">https://doi.org/gfzqhb</a><br />
DOI: <a href="https://doi.org/10.1186/s13104-016-1900-2">10.1186/s13104-016-1900-2</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26868221">26868221</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4751634">PMC4751634</a></p>
</div>
<div id="ref-1AzMUkwox">
<p>51. <strong>Fast and accurate long-read alignment with Burrows–Wheeler transform</strong><br />
Heng Li, Richard Durbin<br />
<em>Bioinformatics</em> (2010-03-01) <a href="https://doi.org/cm27kg">https://doi.org/cm27kg</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btp698">10.1093/bioinformatics/btp698</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/20080505">20080505</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2828108">PMC2828108</a></p>
</div>
<div id="ref-gmH6YDca">
<p>52. <strong>Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM</strong><br />
Heng Li<br />
<em>arXiv</em> (2013-05-28) <a href="https://arxiv.org/abs/1303.3997">https://arxiv.org/abs/1303.3997</a></p>
</div>
<div id="ref-IPc8cy7s">
<p>53. <strong>The Sequence Alignment/Map format and SAMtools</strong><br />
H. Li, B. Handsaker, A. Wysoker, T. Fennell, J. Ruan, N. Homer, G. Marth, G. Abecasis, R. Durbin, 1000 Genome Project Data Processing Subgroup<br />
<em>Bioinformatics</em> (2009-06-08) <a href="https://doi.org/ff6426">https://doi.org/ff6426</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btp352">10.1093/bioinformatics/btp352</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19505943">19505943</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2723002">PMC2723002</a></p>
</div>
<div id="ref-14nSa8zB9">
<p>54. <strong>The Genome Analysis Toolkit: A MapReduce framework for analyzing next-generation DNA sequencing data</strong><br />
A. McKenna, M. Hanna, E. Banks, A. Sivachenko, K. Cibulskis, A. Kernytsky, K. Garimella, D. Altshuler, S. Gabriel, M. Daly, M. A. DePristo<br />
<em>Genome Research</em> (2010-07-19) <a href="https://doi.org/bnzbn6">https://doi.org/bnzbn6</a><br />
DOI: <a href="https://doi.org/10.1101/gr.107524.110">10.1101/gr.107524.110</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/20644199">20644199</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2928508">PMC2928508</a></p>
</div>
<div id="ref-17zUqUTNA">
<p>55. <strong>Predicting the molecular complexity of sequencing libraries</strong><br />
Timothy Daley, Andrew D Smith<br />
<em>Nature Methods</em> (2013-02-24) <a href="https://doi.org/gfx6f5">https://doi.org/gfx6f5</a><br />
DOI: <a href="https://doi.org/10.1038/nmeth.2375">10.1038/nmeth.2375</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23435259">23435259</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3612374">PMC3612374</a></p>
</div>
<div id="ref-O3E8cbBW">
<p>56. <strong>Qualimap 2: advanced multi-sample quality control for high-throughput sequencing data</strong><br />
Konstantin Okonechnikov, Ana Conesa, Fernando García-Alcalde<br />
<em>Bioinformatics</em> (2015-10-01) <a href="https://doi.org/ggxrmx">https://doi.org/ggxrmx</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btv566">10.1093/bioinformatics/btv566</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26428292">26428292</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4708105">PMC4708105</a></p>
</div>
<div id="ref-1Ab7DZE7s">
<p>57. <strong>Integrative-Transcriptomics/DamageProfiler: DamageProfiler v0.4.9</strong><br />
Judith Neukamm, Alexander Peltzer<br />
<em>Zenodo</em> (2019-11-29) <a href="https://doi.org/ggxrmz">https://doi.org/ggxrmz</a><br />
DOI: <a href="https://doi.org/10.5281/zenodo.3557708">10.5281/zenodo.3557708</a></p>
</div>
<div id="ref-lh0Dclyl">
<p>58. <strong>Separating endogenous ancient DNA from modern day contamination in a Siberian Neandertal</strong><br />
Pontus Skoglund, Bernd H. Northoff, Michael V. Shunkov, Anatoli P. Derevianko, Svante Pääbo, Johannes Krause, Mattias Jakobsson<br />
<em>Proceedings of the National Academy of Sciences</em> (2014-02-11) <a href="https://doi.org/f2z5sw">https://doi.org/f2z5sw</a><br />
DOI: <a href="https://doi.org/10.1073/pnas.1318934111">10.1073/pnas.1318934111</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24469802">24469802</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3926038">PMC3926038</a></p>
</div>
<div id="ref-18bWeJ11I">
<p>59. <strong>An efficient and scalable analysis framework for variant extraction and refinement from population-scale DNA sequence data</strong><br />
Goo Jun, Mary Kate Wing, Gonçalo R. Abecasis, Hyun Min Kang<br />
<em>Genome Research</em> (2015-06) <a href="https://doi.org/f7dz2d">https://doi.org/f7dz2d</a><br />
DOI: <a href="https://doi.org/10.1101/gr.176552.114">10.1101/gr.176552.114</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25883319">25883319</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4448687">PMC4448687</a></p>
</div>
<div id="ref-FyiP2I2k">
<p>60. <strong>ANGSD: Analysis of Next Generation Sequencing Data</strong><br />
Thorfinn Sand Korneliussen, Anders Albrechtsen, Rasmus Nielsen<br />
<em>BMC Bioinformatics</em> (2014-11-25) <a href="https://doi.org/gb8wpz">https://doi.org/gb8wpz</a><br />
DOI: <a href="https://doi.org/10.1186/s12859-014-0356-4">10.1186/s12859-014-0356-4</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25420514">25420514</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4248462">PMC4248462</a></p>
</div>
<div id="ref-sYguBb3Q">
<p>61. <strong>Bioconda: sustainable and comprehensive software distribution for the life sciences</strong><br />
Björn Grüning, Ryan Dale, Andreas Sjödin, Brad A. Chapman, Jillian Rowe, Christopher H. Tomkins-Tinch, Renan Valieris, Johannes Köster, The Bioconda Team<br />
<em>Nature Methods</em> (2018-07-02) <a href="https://doi.org/gd2xzp">https://doi.org/gd2xzp</a><br />
DOI: <a href="https://doi.org/10.1038/s41592-018-0046-7">10.1038/s41592-018-0046-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29967506">29967506</a></p>
</div>
<div id="ref-qj0Rqpq8">
<p>62. <strong>conda-forge | community driven packaging for conda</strong><a href="https://conda-forge.org/">https://conda-forge.org/</a></p>
</div>
<div id="ref-5S7TlWlb">
<p>63. <strong>Schmutzi: estimation of contamination and endogenous mitochondrial consensus calling for ancient DNA</strong><br />
Gabriel Renaud, Viviane Slon, Ana T. Duggan, Janet Kelso<br />
<em>Genome Biology</em> (2015-10-12) <a href="https://doi.org/f72mvg">https://doi.org/f72mvg</a><br />
DOI: <a href="https://doi.org/10.1186/s13059-015-0776-0">10.1186/s13059-015-0776-0</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26458810">26458810</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4601135">PMC4601135</a></p>
</div>
<div id="ref-PiS0h6Mu">
<p>64. <strong>Fast gapped-read alignment with Bowtie 2</strong><br />
Ben Langmead, Steven L Salzberg<br />
<em>Nature Methods</em> (2012-03-04) <a href="https://doi.org/gd2xzn">https://doi.org/gd2xzn</a><br />
DOI: <a href="https://doi.org/10.1038/nmeth.1923">10.1038/nmeth.1923</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22388286">22388286</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3322381">PMC3322381</a></p>
</div>
<div id="ref-pt9vlq4w">
<p>65. <strong>fastp: an ultra-fast all-in-one FASTQ preprocessor</strong><br />
Shifu Chen, Yanqing Zhou, Yaru Chen, Jia Gu<br />
<em>Bioinformatics</em> (2018-09-01) <a href="https://doi.org/gd9mrb">https://doi.org/gd9mrb</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/bty560">10.1093/bioinformatics/bty560</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30423086">30423086</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6129281">PMC6129281</a></p>
</div>
<div id="ref-or1FajUp">
<p>66. <strong>QC Fail Sequencing » Illumina 2 colour chemistry can overcall high confidence G bases</strong><a href="https://sequencing.qcfail.com/articles/illumina-2-colour-chemistry-can-overcall-high-confidence-g-bases/">https://sequencing.qcfail.com/articles/illumina-2-colour-chemistry-can-overcall-high-confidence-g-bases/</a></p>
</div>
<div id="ref-qKTLLip1">
<p>67. <strong>Haplotype-based variant detection from short-read sequencing</strong><br />
Erik Garrison, Gabor Marth<br />
<em>arXiv</em> (2012-07-24) <a href="https://arxiv.org/abs/1207.3907">https://arxiv.org/abs/1207.3907</a></p>
</div>
<div id="ref-194vI7Tnf">
<p>68. <strong>MALT: Fast alignment and analysis of metagenomic DNA sequence data applied to the Tyrolean Iceman</strong><br />
Alexander Herbig, Frank Maixner, Kirsten I. Bos, Albert Zink, Johannes Krause, Daniel H. Huson<br />
<em>bioRxiv</em> (2016-04-27) <a href="https://doi.org/ggxkk9">https://doi.org/ggxkk9</a><br />
DOI: <a href="https://doi.org/10.1101/050559">10.1101/050559</a></p>
</div>
<div id="ref-qDQBcgzB">
<p>69. <strong>Salmonella enterica genomes from victims of a major sixteenth-century epidemic in Mexico</strong><br />
Åshild J. Vågene, Alexander Herbig, Michael G. Campana, Nelly M. Robles García, Christina Warinner, Susanna Sabin, Maria A. Spyrou, Aida Andrades Valtueña, Daniel Huson, Noreen Tuross, … Johannes Krause<br />
<em>Nature Ecology &amp; Evolution</em> (2018-01-15) <a href="https://doi.org/ggxkk7">https://doi.org/ggxkk7</a><br />
DOI: <a href="https://doi.org/10.1038/s41559-017-0446-6">10.1038/s41559-017-0446-6</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29335577">29335577</a></p>
</div>
<div id="ref-wfp4mBcT">
<p>70. <strong>Improved metagenomic analysis with Kraken 2</strong><br />
Derrick E. Wood, Jennifer Lu, Ben Langmead<br />
<em>Genome Biology</em> (2019-11-28) <a href="https://doi.org/ggfk55">https://doi.org/ggfk55</a><br />
DOI: <a href="https://doi.org/10.1186/s13059-019-1891-0">10.1186/s13059-019-1891-0</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31779668">31779668</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6883579">PMC6883579</a></p>
</div>
<div id="ref-cvMiJCKj">
<p>71. <strong>HOPS: automated detection and authentication of pathogen DNA in archaeological remains</strong><br />
Ron Hübler, Felix M. Key, Christina Warinner, Kirsten I. Bos, Johannes Krause, Alexander Herbig<br />
<em>Genome Biology</em> (2019-12-16) <a href="https://doi.org/ggxkmb">https://doi.org/ggxkmb</a><br />
DOI: <a href="https://doi.org/10.1186/s13059-019-1903-0">10.1186/s13059-019-1903-0</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31842945">31842945</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6913047">PMC6913047</a></p>
</div>
<div id="ref-BF7nFrAi">
<p>72. <strong>Microbial differences between dental plaque and historic dental calculus are related to oral biofilm maturation stage</strong><br />
Irina M. Velsko, James A. Fellows Yates, Franziska Aron, Richard W. Hagan, Laurent A. F. Frantz, Louise Loe, Juan Bautista Rodriguez Martinez, Eros Chaves, Chris Gosden, Greger Larson, Christina Warinner<br />
<em>Microbiome</em> (2019-07-06) <a href="https://doi.org/ggxkmc">https://doi.org/ggxkmc</a><br />
DOI: <a href="https://doi.org/10.1186/s40168-019-0717-3">10.1186/s40168-019-0717-3</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31279340">31279340</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6612086">PMC6612086</a></p>
</div>
<div id="ref-1HWiAHnIw">
<p>73. <strong>BEDTools: a flexible suite of utilities for comparing genomic features</strong><br />
Aaron R. Quinlan, Ira M. Hall<br />
<em>Bioinformatics</em> (2010-03-15) <a href="https://doi.org/cmrms3">https://doi.org/cmrms3</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btq033">10.1093/bioinformatics/btq033</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/20110278">20110278</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2832824">PMC2832824</a></p>
</div>
<div id="ref-KlTB2lhf">
<p>74. <strong>Ancient Fennoscandian genomes reveal origin and spread of Siberian ancestry in Europe</strong><br />
Thiseas C. Lamnidis, Kerttu Majander, Choongwon Jeong, Elina Salmela, Anna Wessman, Vyacheslav Moiseyev, Valery Khartanovich, Oleg Balanovsky, Matthias Ongyerth, Antje Weihmann, … Stephan Schiffels<br />
<em>Nature Communications</em> (2018-11-27) <a href="https://doi.org/ggxkk6">https://doi.org/ggxkk6</a><br />
DOI: <a href="https://doi.org/10.1038/s41467-018-07483-5">10.1038/s41467-018-07483-5</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30479341">30479341</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6258758">PMC6258758</a></p>
</div>
<div id="ref-U9WbDW7I">
<p>75. <strong>MultiQC: summarize analysis results for multiple tools and samples in a single report</strong><br />
Philip Ewels, Måns Magnusson, Sverker Lundin, Max Käller<br />
<em>Bioinformatics</em> (2016-10-01) <a href="https://doi.org/f3s996">https://doi.org/f3s996</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btw354">10.1093/bioinformatics/btw354</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27312411">27312411</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5039924">PMC5039924</a></p>
</div>
<div id="ref-NXbAmDI0">
<p>76. <strong>nf-core/eager</strong><br />
nf-core<br />
(2020-06-03) <a href="https://github.com/nf-core/eager">https://github.com/nf-core/eager</a></p>
</div>
<div id="ref-13kS0x2y1">
<p>77. <strong>nf-core/configs</strong><br />
nf-core<br />
(2020-06-05) <a href="https://github.com/nf-core/configs">https://github.com/nf-core/configs</a></p>
</div>
<div id="ref-YuJbg3zO">
<p>78. <strong>Open collaborative writing with Manubot</strong><br />
Daniel S. Himmelstein, Vincent Rubinetti, David R. Slochower, Dongbo Hu, Venkat S. Malladi, Casey S. Greene, Anthony Gitter<br />
<em>PLOS Computational Biology</em> (2019-06-24) <a href="https://doi.org/c7np">https://doi.org/c7np</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pcbi.1007128">10.1371/journal.pcbi.1007128</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31233491">31233491</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6611653">PMC6611653</a></p>
</div>
</div>
<!-- default theme -->

<style>
    /* import google fonts */
    @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
    @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

    /* -------------------------------------------------- */
    /* global */
    /* -------------------------------------------------- */

    /* all elements */
    * {
        /* force sans-serif font unless specified otherwise */
        font-family: "Open Sans", "Helvetica", sans-serif;

        /* prevent text inflation on some mobile browsers */
        -webkit-text-size-adjust: none !important;
        -moz-text-size-adjust: none !important;
        -o-text-size-adjust: none !important;
        text-size-adjust: none !important;
    }

    @media only screen {
        /* "page" element */
        body {
            position: relative;
            box-sizing: border-box;
            font-size: 12pt;
            line-height: 1.5;
            max-width: 8.5in;
            margin: 20px auto;
            padding: 40px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* "page" element */
        body {
            padding: 20px;
            margin: 0;
            border-radius: 0;
            border: none;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
            background: none;
        }
    }

    /* -------------------------------------------------- */
    /* headings */
    /* -------------------------------------------------- */

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
        margin: 20px 0;
        padding: 0;
        font-weight: bold;
    }

    /* biggest heading */
    h1 {
        margin: 40px 0;
        text-align: center;
    }

    /* second biggest heading */
    h2 {
        margin-top: 30px;
        padding-bottom: 5px;
        border-bottom: solid 1px #bdbdbd;
    }

    /* -------------------------------------------------- */
    /* manuscript header */
    /* -------------------------------------------------- */

    /* manuscript title */
    header > h1 {
        margin: 0;
    }

    /* manuscript title caption text (ie "automatically generated on") */
    header + p {
        text-align: center;
        margin-top: 10px;
    }

    /* -------------------------------------------------- */
    /* text elements */
    /* -------------------------------------------------- */

    /* links */
    a {
        color: #2196f3;
        overflow-wrap: break-word;
    }

    /* normal links (not empty, not button link, not syntax highlighting link) */
    a:not(:empty):not(.button):not(.sourceLine) {
        padding-left: 1px;
        padding-right: 1px;
    }

    /* superscripts and subscripts */
    sub,
    sup {
        /* prevent from affecting line height */
        line-height: 0;
    }

    /* unordered and ordered lists*/
    ul,
    ol {
        padding-left: 20px;
    }

    /* class for styling text semibold */
    .semibold {
        font-weight: 600;
    }

    /* class for styling elements horizontally left aligned */
    .left {
        display: block;
        text-align: left;
        margin-left: auto;
        margin-right: 0;
        justify-content: left;
    }

    /* class for styling elements horizontally centered */
    .center {
        display: block;
        text-align: center;
        margin-left: auto;
        margin-right: auto;
        justify-content: center;
    }

    /* class for styling elements horizontally right aligned */
    .right {
        display: block;
        text-align: right;
        margin-left: 0;
        margin-right: auto;
        justify-content: right;
    }

    /* -------------------------------------------------- */
    /* section elements */
    /* -------------------------------------------------- */

    /* horizontal divider line */
    hr {
        border: none;
        height: 1px;
        background: #bdbdbd;
    }

    /* paragraphs, horizontal dividers, figures, tables, code */
    p,
    hr,
    figure,
    table,
    pre {
        /* treat all as "paragraphs", with consistent vertical margins */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* figures */
    /* -------------------------------------------------- */

    /* figure */
    figure {
        max-width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure caption */
    figcaption {
        padding: 0;
        padding-top: 10px;
    }

    /* figure image element */
    figure img {
        max-width: 100%;
        display: block;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure auto-number */
    img + figcaption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* tables */
    /* -------------------------------------------------- */

    /* table */
    table {
        border-collapse: collapse;
        border-spacing: 0;
        width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* table cells */
    th,
    td {
        border: solid 1px #bdbdbd;
        padding: 10px;
        /* squash table if too wide for page by forcing line breaks */
        overflow-wrap: break-word;
        word-break: break-word;
    }

    /* header row and even rows */
    th,
    tr:nth-child(2n) {
        background-color: #fafafa;
    }

    /* odd rows */
    tr:nth-child(2n + 1) {
        background-color: #ffffff;
    }

    /* table caption */
    caption {
        text-align: left;
        padding: 0;
        padding-bottom: 10px;
    }

    /* table auto-number */
    table > caption > span:first-of-type,
    div.table_wrapper > table > caption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* code */
    /* -------------------------------------------------- */

    /* multi-line code block */
    pre {
        padding: 10px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
        break-inside: avoid;
        text-align: left;
    }

    /* inline code, ie code within normal text */
    :not(pre) > code {
        padding: 0 4px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
    }

    /* code text */
    /* apply all children, to reach syntax highlighting sub-elements */
    code,
    code * {
        /* force monospace font */
        font-family: "Source Code Pro", "Courier New", monospace;
    }

    /* -------------------------------------------------- */
    /* quotes */
    /* -------------------------------------------------- */

    /* quoted text */
    blockquote {
        margin: 0;
        padding: 0;
        border-left: 4px solid #bdbdbd;
        padding-left: 16px;
        break-inside: avoid;
    }

    /* -------------------------------------------------- */
    /* banners */
    /* -------------------------------------------------- */

    /* info banners */
    .banner {
        box-sizing: border-box;
        display: block;
        position: relative;
        width: 100%;
        margin-top: 20px;
        margin-bottom: 20px;
        padding: 20px;
        text-align: center;
    }

    /* paragraph in banner */
    .banner > p {
        margin: 0;
    }

    /* -------------------------------------------------- */
    /* highlight colors */
    /* -------------------------------------------------- */

    .white {
        background: #ffffff;
    }
    .lightgrey {
        background: #eeeeee;
    }
    .grey {
        background: #757575;
    }
    .darkgrey {
        background: #424242;
    }
    .black {
        background: #000000;
    }
    .lightred {
        background: #ffcdd2;
    }
    .lightyellow {
        background: #ffecb3;
    }
    .lightgreen {
        background: #dcedc8;
    }
    .lightblue {
        background: #e3f2fd;
    }
    .lightpurple {
        background: #f3e5f5;
    }
    .red {
        background: #f44336;
    }
    .orange {
        background: #ff9800;
    }
    .yellow {
        background: #ffeb3b;
    }
    .green {
        background: #4caf50;
    }
    .blue {
        background: #2196f3;
    }
    .purple {
        background: #9c27b0;
    }
    .white,
    .lightgrey,
    .lightred,
    .lightyellow,
    .lightgreen,
    .lightblue,
    .lightpurple,
    .orange,
    .yellow,
    .white a,
    .lightgrey a,
    .lightred a,
    .lightyellow a,
    .lightgreen a,
    .lightblue a,
    .lightpurple a,
    .orange a,
    .yellow a {
        color: #000000;
    }
    .grey,
    .darkgrey,
    .black,
    .red,
    .green,
    .blue,
    .purple,
    .grey a,
    .darkgrey a,
    .black a,
    .red a,
    .green a,
    .blue a,
    .purple a {
        color: #ffffff;
    }

    /* -------------------------------------------------- */
    /* buttons */
    /* -------------------------------------------------- */

    /* class for styling links like buttons */
    .button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        margin: 5px;
        padding: 10px 20px;
        font-size: 0.75em;
        font-weight: 600;
        text-transform: uppercase;
        text-decoration: none;
        letter-spacing: 1px;
        background: none;
        color: #2196f3;
        border: solid 1px #bdbdbd;
        border-radius: 5px;
    }

    /* buttons when hovered */
    .button:hover:not([disabled]),
    .icon_button:hover:not([disabled]) {
        cursor: pointer;
        background: #f5f5f5;
    }

    /* buttons when disabled */
    .button[disabled],
    .icon_button[disabled] {
        opacity: 0.35;
        pointer-events: none;
    }

    /* class for styling buttons containg only single icon */
    .icon_button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        text-decoration: none;
        margin: 0;
        padding: 0;
        background: none;
        border-radius: 5px;
        border: none;
        width: 20px;
        height: 20px;
        min-width: 20px;
        min-height: 20px;
    }

    /* icon button inner svg image */
    .icon_button > svg {
        height: 16px;
    }

    /* -------------------------------------------------- */
    /* icons */
    /* -------------------------------------------------- */

    /* class for styling icons inline with text */
    .inline_icon {
        height: 1em;
        position: relative;
        top: 0.125em;
    }

    /* -------------------------------------------------- */
    /* print control */
    /* -------------------------------------------------- */

    @media print {
        @page {
            /* suggested printing margin */
            margin: 0.5in;
        }

        /* document and "page" elements */
        html, body {
            margin: 0;
            padding: 0;
            width: 100%;
            height: 100%;
        }

        /* "page" element */
        body {
            font-size: 11pt !important;
            line-height: 1.35;
        }

        /* all headings */
        h1,
        h2,
        h3,
        h4,
        h5,
        h6 {
            margin: 15px 0;
        }

        /* heading 1 */
        h1 {
            font-size: 1.75em;
        }

        /* heading 2 */
        h2 {
            font-size: 1.25em;
            margin-top: 0;
        }

        /* heading 3 */
        h3 {
            font-size: 1.10em;
        }

        /* figures and tables */
        figure, table {
            font-size: 0.85em;
        }

        /* table cells */
        th,
        td {
            padding: 5px;
        }

        /* shrink font awesome icons */
        i.fas,
        i.fab,
        i.far,
        i.fal {
            transform: scale(0.85);
        }

        /* decrease banner margins */
        .banner {
            margin-top: 15px;
            margin-bottom: 15px;
            padding: 15px;
        }

        /* class for centering an element vertically on its own page */
        .page_center {
            margin: auto;
            width: 100%;
            height: 100%;
            display: flex;
            align-items: center;
            vertical-align: middle;
            break-before: page;
            break-after: page;
        }

        /* always insert a page break before the element */
        .page_break_before {
            break-before: page;
        }

        /* always insert a page break after the element */
        .page_break_after {
            break-after: page;
        }

        /* avoid page break before the element */
        .page_break_before_avoid {
            break-before: avoid;
        }

        /* avoid page break after the element */
        .page_break_after_avoid {
            break-after: avoid;
        }

        /* avoid page break inside the element */
        .page_break_inside_avoid {
            break-inside: avoid;
        }
    }

    /* -------------------------------------------------- */
    /* override pandoc css quirks */
    /* -------------------------------------------------- */

    .sourceCode {
        /* prevent unsightly overflow in wide code blocks */
        overflow: auto !important;
    }

    div.sourceCode {
        /* prevent background fill on top-most code block  container */
        background: none !important;
    }

    .sourceCode * {
        /* force consistent line spacing */
        line-height: 1.5 !important;
    }

    div.sourceCode {
        /* style code block margins same as <pre> element */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* tablenos */
    /* -------------------------------------------------- */

    /* tablenos wrapper */
    .tablenos {
        /* show scrollbar on tables if necessary to prevent overflow */
        width: 100%;
        margin: 20px 0;
    }

    .tablenos > table {
        /* move margins from table to table_wrapper to allow margin collapsing */
        margin: 0;
    }

    @media only screen {
        /* tablenos wrapper */
        .tablenos {
            /* show scrollbar on tables if necessary to prevent overflow */
            overflow-x: auto !important;
        }

        .tablenos th,
        .tablenos td {
            overflow-wrap: unset !important;
            word-break: unset !important;
        }

        /* table in wrapper */
        .tablenos table,
        .tablenos table * {
            /* don't break table words */
            overflow-wrap: normal !important;
        }
    }

    /* -------------------------------------------------- */
    /* mathjax */
    /* -------------------------------------------------- */

    /* mathjax containers */
    .math.display > span:not(.MathJax_Preview) {
        /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
        display: flex !important;
        overflow-x: auto !important;
        overflow-y: hidden !important;
        justify-content: center;
        align-items: center;
        margin: 0 !important;
    }

    /* right click menu */
    .MathJax_Menu {
        border-radius: 5px !important;
        border: solid 1px #bdbdbd !important;
        box-shadow: none !important;
    }

    /* equation auto-number */
    span[id^="eq:"] > span.math.display + span {
        font-weight: 600;
    }

    /* equation */
    span[id^="eq:"] > span.math.display > span {
        /* nudge to make room for equation auto-number and anchor */
        margin-right: 60px !important;
    }

    /* -------------------------------------------------- */
    /* anchors plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anchor button */
        .anchor {
            opacity: 0;
            margin-left: 5px;
        }

        /* anchor buttons within <h2>'s */
        h2 .anchor {
            margin-left: 10px;
        }

        /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
        *:hover > .anchor,
        .anchor:hover,
        .anchor:focus {
            opacity: 1;
        }

        /* anchor button when hovered */
        .anchor:hover {
            cursor: pointer;
        }
    }

    /* always show anchor button on devices with no mouse/hover ability */
    @media (hover: none) {
        .anchor {
            opacity: 1;
        }
    }

    /* always hide anchor button on print */
    @media only print {
        .anchor {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* accordion plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* accordion arrow button */
        .accordion_arrow {
            margin-right: 10px;
        }

        /* arrow icon when <h2> data-collapsed attribute true */
        h2[data-collapsed="true"] > .accordion_arrow > svg {
            transform: rotate(-90deg);
        }

        /* all elements (except <h2>'s) when data-collapsed attribute true */
        *:not(h2)[data-collapsed="true"] {
            display: none;
        }

        /* accordion arrow button when hovered and <h2>'s when hovered */
        .accordion_arrow:hover,
        h2[data-collapsed="true"]:hover,
        h2[data-collapsed="false"]:hover {
            cursor: pointer;
        }
    }

    /* always hide accordion arrow button on print */
    @media only print {
        .accordion_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* tooltips plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* tooltip container */
        #tooltip {
            position: absolute;
            width: 50%;
            min-width: 240px;
            max-width: 75%;
            z-index: 1;
        }

        /* tooltip content */
        #tooltip_content {
            margin-bottom: 5px;
            padding: 20px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
            overflow-wrap: break-word;
        }

        /* tooltip copy of paragraphs and figures */
        #tooltip_content > p,
        #tooltip_content > figure {
            margin: 0;
            max-height: 320px;
            overflow-y: auto;
        }

        /* tooltip copy of <img> */
        #tooltip_content > figure > img {
            max-height: 260px;
        }

        /* navigation bar */
        #tooltip_nav_bar {
            margin-top: 10px;
            text-align: center;
        }

        /* navigation bar previous/next buton */
        #tooltip_nav_bar > .icon_button {
            position: relative;
            top: 3px;
        }

        /* navigation bar previous button */
        #tooltip_nav_bar > .icon_button:first-of-type {
            margin-right: 5px;
        }

        /* navigation bar next button */
        #tooltip_nav_bar > .icon_button:last-of-type {
            margin-left: 5px;
        }
    }

    /* always hide tooltip on print */
    @media only print {
        #tooltip {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* jump to first plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* jump button */
        .jump_arrow {
            position: relative;
            top: 0.125em;
            margin-right: 5px;
        }
    }

    /* always hide jump button on print */
    @media only print {
        .jump_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* link highlight plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anything with data-highlighted attribute true */
        [data-highlighted="true"] {
            background: #ffeb3b;
        }

        /* anything with data-selected attribute true */
        [data-selected="true"] {
            background: #ff8a65 !important;
        }

        /* animation definition for glow */
        @keyframes highlight_glow {
            0% {
                background: none;
            }
            10% {
                background: #bbdefb;
            }
            100% {
                background: none;
            }
        }

        /* anything with data-glow attribute true */
        [data-glow="true"] {
            animation: highlight_glow 2s;
        }
    }

    /* -------------------------------------------------- */
    /* table of contents plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* toc panel */
        #toc_panel {
            box-sizing: border-box;
            position: fixed;
            top: 0;
            left: 0;
            background: #ffffff;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            z-index: 2;
        }

        /* toc panel when closed */
        #toc_panel[data-open="false"] {
            min-width: 60px;
            width: 60px;
            height: 60px;
            border-right: solid 1px #bdbdbd;
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc panel when open */
        #toc_panel[data-open="true"] {
            min-width: 260px;
            max-width: 480px;
            /* keep panel edge consistent distance away from "page" edge */
            width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
            bottom: 0;
            border-right: solid 1px #bdbdbd;
        }

        /* toc panel header */
        #toc_header {
            box-sizing: border-box;
            display: flex;
            flex-direction: row;
            align-items: center;
            height: 60px;
            margin: 0;
            padding: 20px;
        }

        /* toc panel header when hovered */
        #toc_header:hover {
            cursor: pointer;
        }

        /* toc panel header when panel open */
        #toc_panel[data-open="true"] > #toc_header {
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc open/close header button */
        #toc_button {
            margin-right: 20px;
        }

        /* hide toc list and header text when closed */
        #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
        #toc_panel[data-open="false"] > #toc_list {
            display: none;
        }

        /* toc list of entries */
        #toc_list {
            box-sizing: border-box;
            width: 100%;
            padding: 20px;
            position: absolute;
            top: calc(60px + 1px);
            bottom: 0;
            overflow: auto;
        }

        /* toc entry, link to section in document */
        .toc_link {
            display: block;
            padding: 5px;
            position: relative;
            font-weight: 600;
            text-decoration: none;
        }

        /* toc entry when hovered or when "viewed" */
        .toc_link:hover,
        .toc_link[data-viewing="true"] {
            background: #f5f5f5;
        }

        /* toc entry, level 1 indentation */
        .toc_link[data-level="1"] {
            margin-left: 0;
        }

        /* toc entry, level 2 indentation */
        .toc_link[data-level="2"] {
            margin-left: 20px;
        }

        /* toc entry, level 3 indentation */
        .toc_link[data-level="3"] {
            margin-left: 40px;
        }

        /* toc entry, level 4 indentation */
        .toc_link[data-level="4"] {
            margin-left: 60px;
        }

        /* toc entry bullets */
        #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
            position: absolute;
            left: -15px;
            top: -1px;
            font-size: 1.5em;
        }

        /* toc entry, level 2 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
            content: "\2022";
        }

        /* toc entry, level 3 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
            content: "\25AB";
        }

        /* toc entry, level 4 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
            content: "-";
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* push <body> ("page") element down to make room for toc icon */
        .toc_body_nudge {
            padding-top: 60px;
        }

        /* toc icon when panel closed and not hovered */
        #toc_panel[data-open="false"]:not(:hover) {
            background: rgba(255, 255, 255, 0.75);
        }
    }

    /* always hide toc panel on print */
    @media only print {
        #toc_panel {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* lightbox plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* regular <img> in document when hovered */
        .lightbox_document_img:hover {
            cursor: pointer;
        }

        .body_no_scroll {
            overflow: hidden !important;
        }

        /* screen overlay */
        #lightbox_overlay {
            display: flex;
            flex-direction: column;
            position: fixed;
            left: 0;
            top: 0;
            right: 0;
            bottom: 0;
            background: rgba(0, 0, 0, 0.75);
            z-index: 3;
        }

        /* middle area containing lightbox image */
        #lightbox_image_container {
            flex-grow: 1;
            display: flex;
            justify-content: center;
            align-items: center;
            overflow: hidden;
            position: relative;
            padding: 20px;
        }

        /* bottom area containing caption */
        #lightbox_bottom_container {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 100px;
            min-height: 100px;
            max-height: 100px;
            background: rgba(0, 0, 0, 0.5);
        }

        /* image number info text box */
        #lightbox_number_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            left: 2px;
            top: 0;
            z-index: 4;
        }

        /* zoom info text box */
        #lightbox_zoom_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            right: 2px;
            top: 0;
            z-index: 4;
        }

        /* copy of image caption */
        #lightbox_caption {
            box-sizing: border-box;
            display: inline-block;
            width: 100%;
            max-height: 100%;
            padding: 10px 0;
            text-align: center;
            overflow-y: auto;
            color: #ffffff;
        }

        /* navigation previous/next button */
        .lightbox_button {
            width: 100px;
            height: 100%;
            min-width: 100px;
            min-height: 100%;
            color: #ffffff;
        }

        /* navigation previous/next button when hovered */
        .lightbox_button:hover {
            background: none !important;
        }

        /* navigation button icon */
        .lightbox_button > svg {
            height: 25px;
        }

        /* figure auto-number */
        #lightbox_caption > span:first-of-type {
            font-weight: bold;
            margin-right: 5px;
        }

        /* lightbox image when hovered */
        #lightbox_img:hover {
            cursor: grab;
        }

        /* lightbox image when grabbed */
        #lightbox_img:active {
            cursor: grabbing;
        }
    }

    /* when on screen < 480px wide */
    @media only screen and (max-width: 480px) {
        /* make navigation buttons skinnier on small screens to make more room for caption text */
        .lightbox_button {
            width: 50px;
            min-width: 50px;
        }
    }

    /* always hide lightbox on print */
    @media only print {
        #lightbox_overlay {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* hypothesis (annotations) plugin */
    /* -------------------------------------------------- */

    /* side panel */
    .annotator-frame {
        width: 280px !important;
        z-index: 0 !important;
    }

    /* match highlight color to rest of theme */
    .annotator-highlights-always-on .annotator-hl {
        background-color: #ffeb3b !important;
    }

    /* match focused color to rest of theme */
    .annotator-hl.annotator-hl-focused {
        background-color: #ff8a65 !important;
    }

    /* match bucket bar color to rest of theme */
    .annotator-bucket-bar {
        background: #f5f5f5 !important;
    }

    /* always hide toolbar and tooltip on print */
    @media only print {
        .annotator-frame {
            display: none !important;
        }

        hypothesis-adder {
            display: none !important;
        }
    }
</style>
<!-- anchors plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds an anchor next to each of a certain type
        // of element that provides a human-readable url to that specific
        // item/position in the document (eg "manuscript.html#abstract"). It
        // also makes it such that scrolling out of view of a target removes
        // its identifier from the url.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'anchors';

        // default plugin options
        const options = {
            // which types of elements to add anchors next to, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3, [id^="fig:"], [id^="tbl:"], [id^="eq:"]',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // add anchor to each element of specified types
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements)
                addAnchor(element);

            // attach scroll listener to window
            window.addEventListener('scroll', onScroll);
        }

        // when window is scrolled
        function onScroll() {
            // if url has hash and user has scrolled out of view of hash
            // target, remove hash from url
            const tolerance = 100;
            const target = getHashTarget();
            if (target) {
                if (
                    target.getBoundingClientRect().top >
                        window.innerHeight + tolerance ||
                    target.getBoundingClientRect().bottom < 0 - tolerance
                )
                    history.pushState(null, null, ' ');
            }
        }

        // add anchor to element
        function addAnchor(element) {
            let addTo; // element to add anchor button to

            // if figure or table, modify withId and addTo to get expected
            // elements
            if (element.id.indexOf('fig:') === 0) {
                addTo = element.querySelector('figcaption');
            } else if (element.id.indexOf('tbl:') === 0) {
                addTo = element.querySelector('caption');
            } else if (element.id.indexOf('eq:') === 0) {
                addTo = element.querySelector('.eqnos-number');
            }

            addTo = addTo || element;
            const id = element.id || null;

            // do not add anchor if element doesn't have assigned id.
            // id is generated by pandoc and is assumed to be unique and
            // human-readable
            if (!id)
                return;

            // create anchor button
            const anchor = document.createElement('a');
            anchor.innerHTML = document.querySelector('.icon_link').innerHTML;
            anchor.title = 'Link to this part of the document';
            anchor.classList.add('icon_button', 'anchor');
            anchor.dataset.ignore = 'true';
            anchor.href = '#' + id;
            addTo.appendChild(anchor);
        }

        // get element that is target of link or url hash
        function getHashTarget() {
            const hash = window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');
            if (id.indexOf('tbl:') === 0)
                target = target.querySelector('table');

            return target;
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- link icon -->

<template class="icon_link">
    <!-- modified from: https://fontawesome.com/icons/link -->
    <svg width="16" height="16" viewBox="0 0 512 512">
        <path
            fill="currentColor"
            d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
        ></path>
    </svg>
</template>
<!-- accordion plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows sections of content under <h2> headings
        // to be collapsible.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'accordion';

        // default plugin options
        const options = {
            // whether to always start expanded ('false'), always start
            // collapsed ('true'), or start collapsed when screen small ('auto')
            startCollapsed: 'auto',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <h2> heading
            const headings = document.querySelectorAll('h2');
            for (const heading of headings) {
                addArrow(heading);

                // start expanded/collapsed based on option
                if (
                    options.startCollapsed === 'true' ||
                    (options.startCollapsed === 'auto' && isSmallScreen())
                )
                    collapseHeading(heading);
                else
                    expandHeading(heading);
            }

            // attach hash change listener to window
            window.addEventListener('hashchange', onHashChange);
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                goToElement(target);
        }

        // add arrow to heading
        function addArrow(heading) {
            // add arrow button
            const arrow = document.createElement('button');
            arrow.innerHTML = document.querySelector(
                '.icon_angle_down'
            ).innerHTML;
            arrow.classList.add('icon_button', 'accordion_arrow');
            heading.insertBefore(arrow, heading.firstChild);

            // attach click listener to heading and button
            heading.addEventListener('click', onHeadingClick);
            arrow.addEventListener('click', onArrowClick);
        }

        // determine if on mobile-like device with small screen
        function isSmallScreen() {
            return Math.min(window.innerWidth, window.innerHeight) < 480;
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get element that is target of hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');
            if (id.indexOf('tbl:') === 0)
                target = target.querySelector('table');

            return target;
        }

        // when <h2> heading is clicked
        function onHeadingClick(event) {
            // only collapse if <h2> itself is target of click (eg, user did
            // not click on anchor within <h2>)
            if (event.target === this)
                toggleCollapse(this);
        }

        // when arrow button is clicked
        function onArrowClick() {
            toggleCollapse(this.parentNode);
        }

        // collapse section if expanded, expand if collapsed
        function toggleCollapse(heading) {
            if (heading.dataset.collapsed === 'false')
                collapseHeading(heading);
            else
                expandHeading(heading);
        }

        // elements to exclude from collapse, such as table of contents panel,
        // hypothesis panel, etc
        const exclude = '#toc_panel, div.annotator-frame, #lightbox_overlay';

        // collapse section
        function collapseHeading(heading) {
            heading.setAttribute('data-collapsed', 'true');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'true');
        }

        // expand section
        function expandHeading(heading) {
            heading.setAttribute('data-collapsed', 'false');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'false');
        }

        // get list of elements between this <h2> and next <h2> or <h1>
        // ("children" of the <h2> section)
        function getChildren(heading) {
            return nextUntil(heading, 'h2, h1', exclude);
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get list of elements after a start element up to element matching
        // query
        function nextUntil(element, query, exclude) {
            const elements = [];
            while (element = element.nextElementSibling, element) {
                if (element.matches(query))
                    break;
                if (!element.matches(exclude))
                    elements.push(element);
            }
            return elements;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
    <!-- modified from: https://fontawesome.com/icons/angle-down -->
    <svg width="16" height="16" viewBox="0 0 448 512">
        <path
            fill="currentColor"
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        ></path>
    </svg>
</template>
<!-- tooltips plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when the user hovers or
        // focuses a link to a citation or figure, a tooltip appears with a
        // preview of the reference content, along with arrows to navigate
        // between instances of the same reference in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tooltips';

        // default plugin options
        const options = {
            // whether user must click off to close tooltip instead of just
            // un-hovering
            clickClose: 'false',
            // delay (in ms) between opening and closing tooltip
            delay: '100',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach hover and focus listeners to link
                link.addEventListener('mouseover', onLinkHover);
                link.addEventListener('mouseleave', onLinkUnhover);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('touchend', onLinkTouch);
            }

            // attach mouse, key, and resize listeners to window
            window.addEventListener('mousedown', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('keyup', onKeyUp);
            window.addEventListener('resize', onResize);
        }

        // when link is hovered
        function onLinkHover() {
            // function to open tooltip
            const delayOpenTooltip = function() {
                openTooltip(this);
            }.bind(this);

            // run open function after delay
            this.openTooltipTimer = window.setTimeout(
                delayOpenTooltip,
                options.delay
            );
        }

        // when mouse leaves link
        function onLinkUnhover() {
            // cancel opening tooltip
            window.clearTimeout(this.openTooltipTimer);

            // don't close on unhover if option specifies
            if (options.clickClose === 'true')
                return;

            // function to close tooltip
            const delayCloseTooltip = function() {
                // if tooltip open and if mouse isn't over tooltip, close
                const tooltip = document.getElementById('tooltip');
                if (tooltip && !tooltip.matches(':hover'))
                    closeTooltip();
            };

            // run close function after delay
            this.closeTooltipTimer = window.setTimeout(
                delayCloseTooltip,
                options.delay
            );
        }

        // when link is focused (tabbed to)
        function onLinkFocus(event) {
            openTooltip(this);
        }

        // when link is touched on touch screen
        function onLinkTouch(event) {
            // attempt to force hover state on first tap always, and trigger
            // regular link click (and navigation) on second tap
            if (event.target === document.activeElement)
                event.target.click();
            else {
                document.activeElement.blur();
                event.target.focus();
            }
            if (event.cancelable)
                event.preventDefault();
            event.stopPropagation();
            return false;
        }

        // when mouse is clicked anywhere in window
        function onClick(event) {
            closeTooltip();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'tooltip_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'tooltip_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeTooltip();
                    break;
            }
        }

        // when window is resized or zoomed
        function onResize() {
            closeTooltip();
        }

        // get all links of types we wish to handle
        function getLinks() {
            const queries = [];
            // exclude buttons, anchor links, toc links, etc
            const exclude =
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            queries.push('a[href^="#ref-"]' + exclude); // citation links
            queries.push('a[href^="#fig:"]' + exclude); // figure links
            const query = queries.join(', ');
            return document.querySelectorAll(query);
        }

        // get links with same target, get index of link in set, get total
        // same links
        function getSameLinks(link) {
            const sameLinks = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    sameLinks.push(otherLink);
            }

            return {
                elements: sameLinks,
                index: sameLinks.indexOf(link),
                total: sameLinks.length
            };
        }

        // open tooltip
        function openTooltip(link) {
            // delete tooltip if it exists, start fresh
            closeTooltip();

            // make tooltip element
            const tooltip = makeTooltip(link);

            // if source couldn't be found and tooltip not made, exit
            if (!tooltip)
                return;

            // make navbar elements
            const navBar = makeNavBar(link);
            if (navBar)
                tooltip.firstElementChild.appendChild(navBar);

            // attach tooltip to page
            document.body.appendChild(tooltip);

            // position tooltip
            const position = function() {
                positionTooltip(link);
            };
            position();

            // if tooltip contains images, position again after they've loaded
            const imgs = tooltip.querySelectorAll('img');
            for (const img of imgs)
                img.addEventListener('load', position);
        }

        // close (delete) tooltip
        function closeTooltip() {
            const tooltip = document.getElementById('tooltip');
            if (tooltip)
                tooltip.remove();
        }

        // make tooltip
        function makeTooltip(link) {
            // get target element that link points to
            const source = getSource(link);

            // if source can't be found, exit
            if (!source)
                return;

            // create new tooltip
            const tooltip = document.createElement('div');
            tooltip.id = 'tooltip';
            const tooltipContent = document.createElement('div');
            tooltipContent.id = 'tooltip_content';
            tooltip.appendChild(tooltipContent);

            // make copy of source node and put in tooltip
            const sourceCopy = makeCopy(source);
            tooltipContent.appendChild(sourceCopy);

            // attach mouse event listeners
            tooltip.addEventListener('click', onTooltipClick);
            tooltip.addEventListener('mousedown', onTooltipClick);
            tooltip.addEventListener('touchstart', onTooltipClick);
            tooltip.addEventListener('mouseleave', onTooltipUnhover);

            // (for interaction with lightbox plugin)
            // transfer click on tooltip copied img to original img
            const sourceImg = source.querySelector('img');
            const sourceCopyImg = sourceCopy.querySelector('img');
            if (sourceImg && sourceCopyImg) {
                const clickImg = function() {
                    sourceImg.click();
                    closeTooltip();
                };
                sourceCopyImg.addEventListener('click', clickImg);
            }

            return tooltip;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // when tooltip is clicked
        function onTooltipClick(event) {
            // when user clicks on tooltip, stop click from transferring
            // outside of tooltip (eg, click off to close tooltip, or eg click
            // off to unhighlight same refs)
            event.stopPropagation();
        }

        // when tooltip is unhovered
        function onTooltipUnhover(event) {
            if (options.clickClose === 'true')
                return;

            // make sure new mouse/touch/focus no longer over tooltip or any
            // element within it
            const tooltip = document.getElementById('tooltip');
            if (!tooltip)
                return;
            if (this.contains(event.relatedTarget))
                return;

            closeTooltip();
        }

        // make nav bar to go betwen prev/next instances of same reference
        function makeNavBar(link) {
            // find other links to the same source
            const sameLinks = getSameLinks(link);

            // don't show nav bar when singular reference
            if (sameLinks.total <= 1)
                return;

            // find prev/next links with same target
            const prevLink = getPrevLink(link, sameLinks);
            const nextLink = getNextLink(link, sameLinks);

            // create nav bar
            const navBar = document.createElement('div');
            navBar.id = 'tooltip_nav_bar';
            const text = sameLinks.index + 1 + ' of ' + sameLinks.total;

            // create nav bar prev/next buttons
            const prevButton = document.createElement('button');
            const nextButton = document.createElement('button');
            prevButton.id = 'tooltip_prev_button';
            nextButton.id = 'tooltip_next_button';
            prevButton.title =
                'Jump to the previous occurence of this item in the document [←]';
            nextButton.title =
                'Jump to the next occurence of this item in the document [→]';
            prevButton.classList.add('icon_button');
            nextButton.classList.add('icon_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;
            navBar.appendChild(prevButton);
            navBar.appendChild(document.createTextNode(text));
            navBar.appendChild(nextButton);

            // attach click listeners to buttons
            prevButton.addEventListener('click', function() {
                onPrevNextClick(link, prevLink);
            });
            nextButton.addEventListener('click', function() {
                onPrevNextClick(link, nextLink);
            });

            return navBar;
        }

        // get previous link with same target
        function getPrevLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if < 1
            let index;
            if (sameLinks.index - 1 >= 0)
                index = sameLinks.index - 1;
            else
                index = sameLinks.total - 1;
            return sameLinks.elements[index];
        }

        // get next link with same target
        function getNextLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if > total
            let index;
            if (sameLinks.index + 1 <= sameLinks.total - 1)
                index = sameLinks.index + 1;
            else
                index = 0;
            return sameLinks.elements[index];
        }

        // get element that is target of link or url hash
        function getSource(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if ref or figure, modify target to get expected element
            if (id.indexOf('ref-') === 0)
                target = target.querySelector('p');
            else if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');

            return target;
        }

        // when prev/next arrow button is clicked
        function onPrevNextClick(link, prevNextLink) {
            if (link && prevNextLink)
                goToElement(prevNextLink, window.innerHeight * 0.5);
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // determine position to place tooltip based on link position in
        // viewport and tooltip size
        function positionTooltip(link, left, top) {
            const tooltipElement = document.getElementById('tooltip');
            if (!tooltipElement)
                return;

            // get convenient vars for position/dimensions of
            // link/tooltip/page/view
            link = getRectInPage(link);
            const tooltip = getRectInPage(tooltipElement);
            const view = getRectInPage();

            // horizontal positioning
            if (left)
                // use explicit value
                left = left;
            else if (link.left + tooltip.width < view.right)
                // fit tooltip to right of link
                left = link.left;
            else if (link.right - tooltip.width > view.left)
                // fit tooltip to left of link
                left = link.right - tooltip.width;
            // center tooltip in view
            else
                left = (view.right - view.left) / 2 - tooltip.width / 2;

            // vertical positioning
            if (top)
                // use explicit value
                top = top;
            else if (link.top - tooltip.height > view.top)
                // fit tooltip above link
                top = link.top - tooltip.height;
            else if (link.bottom + tooltip.height < view.bottom)
                // fit tooltip below link
                top = link.bottom;
            else {
                // center tooltip in view
                top = view.top + view.height / 2 - tooltip.height / 2;
                // nudge off of link to left/right if possible
                if (link.right + tooltip.width < view.right)
                    left = link.right;
                else if (link.left - tooltip.width > view.left)
                    left = link.left - tooltip.width;
            }

            tooltipElement.style.left = left + 'px';
            tooltipElement.style.top = top + 'px';
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get position of element relative to page
        function getRectInPage(element) {
            const rect = getRectInView(element);
            const body = getRectInView(document.body);

            const newRect = {};
            newRect.left = rect.left - body.left;
            newRect.top = rect.top - body.top;
            newRect.right = rect.right - body.left;
            newRect.bottom = rect.bottom - body.top;
            newRect.width = rect.width;
            newRect.height = rect.height;

            return newRect;
        }

        // (for interaction with accordion plugin)
        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // (for interaction with accordion plugin)
        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- jump to first plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds a button next to each reference entry,
        // figure, and table that jumps the page to the first occurrence of a
        // link to that item in the manuscript.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'jumpToFirst';

        // default plugin options
        const options = {
            // whether to add buttons next to reference entries
            references: 'true',
            // whether to add buttons next to figures
            figures: 'true',
            // whether to add buttons next to tables
            tables: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            if (options.references !== 'false')
                makeReferenceButtons();
            if (options.figures !== 'false')
                makeFigureButtons();
            if (options.tables !== 'false')
                makeTableButtons();
        }

        // when jump button clicked
        function onButtonClick() {
            const first = getFirstOccurrence(this.dataset.id);
            if (!first)
                return;

            // update url hash so navigating "back" in history will return
            // user to jump button
            window.location.hash = this.dataset.id;
            // scroll to link
            window.setTimeout(function() {
                goToElement(first, window.innerHeight * 0.5);
            }, 0);
        }

        // get first occurence of link to item in document
        function getFirstOccurrence(id) {
            let query = 'a';
            query += '[href="#' + id + '"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelector(query);
        }

        // add button next to each reference entry
        function makeReferenceButtons() {
            const references = document.querySelectorAll('div[id^="ref-"]');
            for (const reference of references) {
                // get reference id and element to add button to
                const id = reference.id;
                const container = reference.firstElementChild;
                const first = getFirstOccurrence(id);

                // if can't find link to reference, ignore
                if (!first)
                    continue;

                // make jump button
                let button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this reference in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.innerHTML = button.outerHTML + container.innerHTML;
                button = container.firstElementChild;
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeFigureButtons() {
            const figures = document.querySelectorAll('[id^="fig:"]');
            for (const figure of figures) {
                // get figure id and element to add button to
                const id = figure.id;
                const container = figure.querySelector('figcaption') || figure;
                const first = getFirstOccurrence(id);

                // if can't find link to figure, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this figure in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeTableButtons() {
            const tables = document.querySelectorAll('[id^="tbl:"]');
            for (const table of tables) {
                // get ref id and element to add button to
                const id = table.id;
                const container = table.querySelector('caption') || table;
                const first = getFirstOccurrence(id);

                // if can't find link to table, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this table in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
    <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
    <svg width="16" height="16" viewBox="0 0 320 512">
        <path
            fill="currentColor"
            d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
        ></path>
    </svg>
</template>
<!-- link highlight plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user hovers or
        // focuses a link, other links that have the same target will be
        // highlighted. It also makes it such that when clicking a link, the
        // target of the link (eg reference, figure, table) is briefly
        // highlighted.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'linkHighlight';

        // default plugin options
        const options = {
            // whether to also highlight links that go to external urls
            externalLinks: 'false',
            // whether user must click off to unhighlight instead of just
            // un-hovering
            clickUnhighlight: 'false',
            // whether to also highlight links that are unique
            highlightUnique: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach mouse and focus listeners to link
                link.addEventListener('mouseenter', onLinkFocus);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('mouseleave', onLinkUnhover);
            }

            // attach click and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('hashchange', onHashChange);

            // run hash change on window load in case user has navigated
            // directly to hash
            onHashChange();
        }

        // when link is focused (tabbed to) or hovered
        function onLinkFocus() {
            highlight(this);
        }

        // when link is unhovered
        function onLinkUnhover() {
            if (options.clickUnhighlight !== 'true')
                unhighlightAll();
        }

        // when the mouse is clicked anywhere in window
        function onClick(event) {
            unhighlightAll();
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                glowElement(target);
        }

        // get element that is target of link or url hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            return target;
        }

        // start glow sequence on an element
        function glowElement(element) {
            const startGlow = function() {
                onGlowEnd();
                element.dataset.glow = 'true';
                element.addEventListener('animationend', onGlowEnd);
            };
            const onGlowEnd = function() {
                element.removeAttribute('data-glow');
                element.removeEventListener('animationend', onGlowEnd);
            };
            startGlow();
        }

        // highlight link and all others with same target
        function highlight(link) {
            // force unhighlight all to start fresh
            unhighlightAll();

            // get links with same target
            if (!link)
                return;
            const sameLinks = getSameLinks(link);

            // if link unique and option is off, exit and don't highlight
            if (sameLinks.length <= 1 && options.highlightUnique !== 'true')
                return;

            // highlight all same links, and "select" (special highlight) this
            // one
            for (const sameLink of sameLinks) {
                if (sameLink === link)
                    sameLink.setAttribute('data-selected', 'true');
                else
                    sameLink.setAttribute('data-highlighted', 'true');
            }
        }

        // unhighlight all links
        function unhighlightAll() {
            const links = getLinks();
            for (const link of links) {
                link.setAttribute('data-selected', 'false');
                link.setAttribute('data-highlighted', 'false');
            }
        }

        // get links with same target
        function getSameLinks(link) {
            const results = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    results.push(otherLink);
            }
            return results;
        }

        // get all links of types we wish to handle
        function getLinks() {
            let query = 'a';
            if (options.externalLinks !== 'true')
                query += '[href^="#"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelectorAll(query);
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- table of contents plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin provides a "table of contents" (toc) panel on
        // the side of the document that allows the user to conveniently
        // navigate between sections of the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tableOfContents';

        // default plugin options
        const options = {
            // which types of elements to add links for, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3',
            // whether default behavior is to be closed ('false'), open
            // ('true'), or only open when screen wide enough to fit panel
            // ('auto'). note: still always starts closed when page loads.
            open: 'auto',
            // if list item is more than this many characters, text will be
            // truncated
            charLimit: '50',
            // whether or not to show bullets next to each toc item
            bullets: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // make toc panel and populate with entries (links to document
            // sections)
            const panel = makePanel();
            if (!panel)
                return;
            makeEntries(panel);
            document.body.insertBefore(panel, document.body.firstChild);

            closePanel();

            // attach click, scroll, and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('scroll', onScroll);
            window.addEventListener('hashchange', onScroll);
            window.addEventListener('keyup', onKeyUp);
            onScroll();

            // add class to push document body down out of way of toc button
            document.body.classList.add('toc_body_nudge');
        }

        // determine if screen wide enough to fit toc panel
        function isSmallScreen() {
            // in default theme:
            // 816px = 8.5in = width of "page" (<body>) element
            // 260px = min width of toc panel (*2 for both sides of <body>)
            return window.innerWidth < 816 + 260 * 2;
        }

        // open/close panel based on option and screen size
        function openOrClosePanel() {
            if (
                options.open === 'true' ||
                (options.open === 'auto' && !isSmallScreen())
            )
                openPanel();
            else
                closePanel();
        }

        // when mouse is clicked anywhere in window
        function onClick() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                openOrClosePanel();
        }

        // when window is scrolled or hash changed
        function onScroll() {
            highlightViewed();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            // close on esc
            if (event.key === 'Escape')
                closePanel();
        }

        // find entry of currently viewed document section in toc and highlight
        function highlightViewed() {
            const firstId = getFirstInView(options.typesQuery);

            // get toc entries (links), unhighlight all, then highlight viewed
            const list = document.getElementById('toc_list');
            if (!firstId || !list)
                return;
            const links = list.querySelectorAll('a');
            for (const link of links)
                link.dataset.viewing = 'false';
            const link = list.querySelector('a[href="#' + firstId + '"]');
            if (!link)
                return;
            link.dataset.viewing = 'true';
        }

        // get first or previous toc listed element in top half of view
        function getFirstInView(query) {
            // get all elements matching query and with id
            const elements = document.querySelectorAll(query);
            const elementsWithIds = [];
            for (const element of elements) {
                if (element.id)
                    elementsWithIds.push(element);
            }


            // get first or previous element in top half of view
            for (let i = 0; i < elementsWithIds.length; i++) {
                const element = elementsWithIds[i];
                const prevElement = elementsWithIds[Math.max(0, i - 1)];
                if (element.getBoundingClientRect().top >= 0) {
                    if (
                        element.getBoundingClientRect().top <
                        window.innerHeight / 2
                    )
                        return element.id;
                    else
                        return prevElement.id;
                }
            }
        }

        // make panel
        function makePanel() {
            // create panel
            const panel = document.createElement('div');
            panel.id = 'toc_panel';
            if (options.bullets === 'true')
                panel.dataset.bullets = 'true';

            // create header
            const header = document.createElement('div');
            header.id = 'toc_header';

            // create toc button
            const button = document.createElement('button');
            button.id = 'toc_button';
            button.innerHTML = document.querySelector(
                '.icon_th_list'
            ).innerHTML;
            button.classList.add('icon_button');

            // create header text
            const text = document.createElement('h3');
            text.innerHTML = 'Table of Contents';

            // create container for toc list
            const list = document.createElement('div');
            list.id = 'toc_list';

            // attach click listeners
            panel.addEventListener('click', onPanelClick);
            header.addEventListener('click', onHeaderClick);
            button.addEventListener('click', onButtonClick);

            // attach elements
            header.appendChild(button);
            header.appendChild(text);
            panel.appendChild(header);
            panel.appendChild(list);

            return panel;
        }

        // create toc entries (links) to each element of the specified types
        function makeEntries(panel) {
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements) {
                // do not add link if element doesn't have assigned id
                if (!element.id)
                    continue;

                // create link/list item
                const link = document.createElement('a');
                link.classList.add('toc_link');
                switch (element.tagName.toLowerCase()) {
                    case 'h1':
                        link.dataset.level = '1';
                        break;
                    case 'h2':
                        link.dataset.level = '2';
                        break;
                    case 'h3':
                        link.dataset.level = '3';
                        break;
                    case 'h4':
                        link.dataset.level = '4';
                        break;
                }
                link.title = element.innerText;
                let text = element.innerText;
                if (text.length > options.charLimit)
                    text = text.slice(0, options.charLimit) + '...';
                link.innerHTML = text;
                link.href = '#' + element.id;
                link.addEventListener('click', onLinkClick);

                // attach link
                panel.querySelector('#toc_list').appendChild(link);
            }
        }

        // when panel is clicked
        function onPanelClick(event) {
            // stop click from propagating to window/document and closing panel
            event.stopPropagation();
        }

        // when header itself is clicked
        function onHeaderClick(event) {
            togglePanel();
        }

        // when button is clicked
        function onButtonClick(event) {
            togglePanel();
            // stop header underneath button from also being clicked
            event.stopPropagation();
        }

        // when link is clicked
        function onLinkClick() {
            openOrClosePanel();
        }

        // open panel if closed, close if opened
        function togglePanel() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                closePanel();
            else
                openPanel();
        }

        // open panel
        function openPanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'true';
        }

        // close panel
        function closePanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'false';
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- th list icon -->

<template class="icon_th_list">
    <!-- modified from: https://fontawesome.com/icons/th-list -->
    <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
        <path
            fill="currentColor"
            d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
            tabindex="-1"
        ></path>
    </svg>
</template>
<!-- lightbox plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user clicks on an
        // image, the image fills the screen and the user can pan/drag/zoom
        // the image and navigate between other images in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'lightbox';

        // default plugin options
        const options = {
            // list of possible zoom/scale factors
            zoomSteps:
                '0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1,' +
                '1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8',
            // whether to fit image to view ('fit'), display at 100% and shrink
            // if necessary ('shrink'), or always display at 100% ('100')
            defaultZoom: 'fit',
            // whether to zoom in/out toward center of view ('true') or mouse
            // ('false')
            centerZoom: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <img> element
            const imgs = document.querySelectorAll('figure > img');
            let count = 1;
            for (const img of imgs) {
                img.classList.add('lightbox_document_img');
                img.dataset.number = count;
                img.dataset.total = imgs.length;
                img.addEventListener('click', openLightbox);
                count++;
            }

            // attach mouse and key listeners to window
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('keyup', onKeyUp);
        }

        // when mouse is moved anywhere in window
        function onWindowMouseMove(event) {
            window.mouseX = event.clientX;
            window.mouseY = event.clientY;
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'lightbox_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'lightbox_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeLightbox();
                    break;
            }
        }

        // open lightbox
        function openLightbox() {
            const lightbox = makeLightbox(this);
            if (!lightbox)
                return;

            blurBody(lightbox);
            document.body.appendChild(lightbox);
        }

        // make lightbox
        function makeLightbox(img) {
            // delete lightbox if it exists, start fresh
            closeLightbox();

            // create screen overlay containing lightbox
            const overlay = document.createElement('div');
            overlay.id = 'lightbox_overlay';

            // create image info boxes
            const numberInfo = document.createElement('div');
            const zoomInfo = document.createElement('div');
            numberInfo.id = 'lightbox_number_info';
            zoomInfo.id = 'lightbox_zoom_info';

            // create container for image
            const imageContainer = document.createElement('div');
            imageContainer.id = 'lightbox_image_container';
            const lightboxImg = makeLightboxImg(
                img,
                imageContainer,
                numberInfo,
                zoomInfo
            );
            imageContainer.appendChild(lightboxImg);

            // create bottom container for caption and navigation buttons
            const bottomContainer = document.createElement('div');
            bottomContainer.id = 'lightbox_bottom_container';
            const caption = makeCaption(img);
            const prevButton = makePrevButton(img);
            const nextButton = makeNextButton(img);
            bottomContainer.appendChild(prevButton);
            bottomContainer.appendChild(caption);
            bottomContainer.appendChild(nextButton);

            // attach top middle and bottom to overlay
            overlay.appendChild(numberInfo);
            overlay.appendChild(zoomInfo);
            overlay.appendChild(imageContainer);
            overlay.appendChild(bottomContainer);

            return overlay;
        }

        // make <img> object that is intuitively draggable and zoomable
        function makeLightboxImg(
            sourceImg,
            container,
            numberInfoBox,
            zoomInfoBox
        ) {
            // create copy of source <img>
            const img = sourceImg.cloneNode(true);
            img.classList.remove('lightbox_document_img');
            img.removeAttribute('id');
            img.removeAttribute('width');
            img.removeAttribute('height');
            img.style.position = 'unset';
            img.style.margin = '0';
            img.style.padding = '0';
            img.style.width = '';
            img.style.height = '';
            img.style.minWidth = '';
            img.style.minHeight = '';
            img.style.maxWidth = '';
            img.style.maxHeight = '';
            img.id = 'lightbox_img';

            // build sorted list of unique zoomSteps, always including a 100%
            let zoomSteps = [];
            const optionsZooms = options.zoomSteps.split(/[^0-9.]/);
            for (const optionZoom of optionsZooms) {
                const newZoom = parseFloat(optionZoom);
                if (newZoom && !zoomSteps.includes(newZoom))
                    zoomSteps.push(newZoom);
            }
            if (!zoomSteps.includes(1))
                zoomSteps.push(1);
            zoomSteps = zoomSteps.sort(function sortNumber(a, b) {
                return a - b;
            });

            // <img> object property variables
            let zoom = 1;
            let translateX = 0;
            let translateY = 0;
            let clickMouseX = undefined;
            let clickMouseY = undefined;
            let clickTranslateX = undefined;
            let clickTranslateY = undefined;

            updateNumberInfo();

            // update image numbers displayed in info box
            function updateNumberInfo() {
                numberInfoBox.innerHTML =
                    sourceImg.dataset.number + ' of ' + sourceImg.dataset.total;
            }

            // update zoom displayed in info box
            function updateZoomInfo() {
                let zoomInfo = zoom * 100;
                if (!Number.isInteger(zoomInfo))
                    zoomInfo = zoomInfo.toFixed(2);
                zoomInfoBox.innerHTML = zoomInfo + '%';
            }

            // move to closest zoom step above current zoom
            const zoomIn = function() {
                for (const zoomStep of zoomSteps) {
                    if (zoomStep > zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                updateTransform();
            };

            // move to closest zoom step above current zoom
            const zoomOut = function() {
                zoomSteps.reverse();
                for (const zoomStep of zoomSteps) {
                    if (zoomStep < zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                zoomSteps.reverse();

                updateTransform();
            };

            // update display of <img> based on scale/translate properties
            const updateTransform = function() {
                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                // get new width/height after scale
                const rect = img.getBoundingClientRect();
                // limit translate
                translateX = Math.max(translateX, -rect.width / 2);
                translateX = Math.min(translateX, rect.width / 2);
                translateY = Math.max(translateY, -rect.height / 2);
                translateY = Math.min(translateY, rect.height / 2);

                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                updateZoomInfo();
            };

            // fit <img> to container
            const fit = function() {
                // no x/y offset, 100% zoom by default
                translateX = 0;
                translateY = 0;
                zoom = 1;

                // widths of <img> and container
                const imgWidth = img.naturalWidth;
                const imgHeight = img.naturalHeight;
                const containerWidth = parseFloat(
                    window.getComputedStyle(container).width
                );
                const containerHeight = parseFloat(
                    window.getComputedStyle(container).height
                );

                // how much zooming is needed to fit <img> to container
                const xRatio = imgWidth / containerWidth;
                const yRatio = imgHeight / containerHeight;
                const maxRatio = Math.max(xRatio, yRatio);
                const newZoom = 1 / maxRatio;

                // fit <img> to container according to option
                if (options.defaultZoom === 'shrink') {
                    if (maxRatio > 1)
                        zoom = newZoom;
                } else if (options.defaultZoom === 'fit')
                    zoom = newZoom;

                updateTransform();
            };

            // when mouse wheel is rolled anywhere in container
            const onContainerWheel = function(event) {
                if (!event)
                    return;

                // let ctrl + mouse wheel to zoom behave as normal
                if (event.ctrlKey)
                    return;

                // prevent normal scroll behavior
                event.preventDefault();
                event.stopPropagation();

                // point around which to scale img
                const viewRect = container.getBoundingClientRect();
                const viewX = (viewRect.left + viewRect.right) / 2;
                const viewY = (viewRect.top + viewRect.bottom) / 2;
                const originX = options.centerZoom === 'true' ? viewX : mouseX;
                const originY = options.centerZoom === 'true' ? viewY : mouseY;

                // get point on image under origin
                const oldRect = img.getBoundingClientRect();
                const oldPercentX = (originX - oldRect.left) / oldRect.width;
                const oldPercentY = (originY - oldRect.top) / oldRect.height;

                // increment/decrement zoom
                if (event.deltaY < 0)
                    zoomIn();
                if (event.deltaY > 0)
                    zoomOut();

                // get offset between previous image point and origin
                const newRect = img.getBoundingClientRect();
                const offsetX =
                    originX - (newRect.left + newRect.width * oldPercentX);
                const offsetY =
                    originY - (newRect.top + newRect.height * oldPercentY);

                // translate image to keep image point under origin
                translateX += offsetX;
                translateY += offsetY;

                // perform translate
                updateTransform();
            };

            // when container is clicked
            function onContainerClick(event) {
                // if container itself is target of click, and not other
                // element above it
                if (event.target === this)
                    closeLightbox();
            }

            // when mouse button is pressed on image
            const onImageMouseDown = function(event) {
                // store original mouse position relative to image
                clickMouseX = window.mouseX;
                clickMouseY = window.mouseY;
                clickTranslateX = translateX;
                clickTranslateY = translateY;
                event.stopPropagation();
                event.preventDefault();
            };

            // when mouse button is released anywhere in window
            const onWindowMouseUp = function(event) {
                // reset original mouse position
                clickMouseX = undefined;
                clickMouseY = undefined;
                clickTranslateX = undefined;
                clickTranslateY = undefined;

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mouseup', onWindowMouseUp);
            };

            // when mouse is moved anywhere in window
            const onWindowMouseMove = function(event) {
                if (
                    clickMouseX === undefined ||
                    clickMouseY === undefined ||
                    clickTranslateX === undefined ||
                    clickTranslateY === undefined
                )
                    return;

                // offset image based on original and current mouse position
                translateX = clickTranslateX + window.mouseX - clickMouseX;
                translateY = clickTranslateY + window.mouseY - clickMouseY;
                updateTransform();
                event.preventDefault();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mousemove', onWindowMouseMove);
            };

            // when window is resized
            const onWindowResize = function(event) {
                fit();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('resize', onWindowResize);
            };

            // attach the necessary event listeners
            img.addEventListener('dblclick', fit);
            img.addEventListener('mousedown', onImageMouseDown);
            container.addEventListener('wheel', onContainerWheel);
            container.addEventListener('mousedown', onContainerClick);
            container.addEventListener('touchstart', onContainerClick);
            window.addEventListener('mouseup', onWindowMouseUp);
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('resize', onWindowResize);

            // run fit() after lightbox atttached to document and <img> Loaded
            // so needed container and img dimensions available
            img.addEventListener('load', fit);

            return img;
        }

        // make caption
        function makeCaption(img) {
            const caption = document.createElement('div');
            caption.id = 'lightbox_caption';
            const captionSource = img.nextElementSibling;
            if (captionSource.tagName.toLowerCase() === 'figcaption') {
                const captionCopy = makeCopy(captionSource);
                caption.innerHTML = captionCopy.innerHTML;
            }

            caption.addEventListener('touchstart', function(event) {
                event.stopPropagation();
            });

            return caption;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // make button to jump to previous image in document
        function makePrevButton(img) {
            const prevButton = document.createElement('button');
            prevButton.id = 'lightbox_prev_button';
            prevButton.title = 'Jump to the previous image in the document [←]';
            prevButton.classList.add('icon_button', 'lightbox_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;

            // attach click listeners to button
            prevButton.addEventListener('click', function() {
                getPrevImg(img).click();
            });

            return prevButton;
        }

        // make button to jump to next image in document
        function makeNextButton(img) {
            const nextButton = document.createElement('button');
            nextButton.id = 'lightbox_next_button';
            nextButton.title = 'Jump to the next image in the document [→]';
            nextButton.classList.add('icon_button', 'lightbox_button');
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;

            // attach click listeners to button
            nextButton.addEventListener('click', function() {
                getNextImg(img).click();
            });

            return nextButton;
        }

        // get previous image in document
        function getPrevImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if < 1
            if (index - 1 >= 0)
                index--;
            else
                index = imgs.length - 1;
            return imgs[index];
        }

        // get next image in document
        function getNextImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if > total
            if (index + 1 <= imgs.length - 1)
                index++;
            else
                index = 0;
            return imgs[index];
        }

        // close lightbox
        function closeLightbox() {
            focusBody();

            const lightbox = document.getElementById('lightbox_overlay');
            if (lightbox)
                lightbox.remove();
        }

        // make all elements behind lightbox non-focusable
        function blurBody(overlay) {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.tabIndex = -1;
            document.body.classList.add('body_no_scroll');
        }

        // make all elements focusable again
        function focusBody() {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.removeAttribute('tabIndex');
            document.body.classList.remove('body_no_scroll');
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- attributes plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows arbitrary HTML attributes to be attached
        // to (almost) any element. Place an HTML comment inside or next to the
        // desired element in the format <!-- $attribute="value" -->

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'attributes';

        // default plugin options
        const options = {
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // get list of comments in document
            const comments = findComments();

            for(const comment of comments)
                if (comment.parentElement)
                    addAttributes(
                        comment.parentElement,
                        comment.nodeValue.trim()
                    );
        }

        // add html attributes to specified element based on string of 
        // html attributes and values
        function addAttributes(element, text) {
            // regex's for finding attribute/value pairs in the format of
            // attribute="value" or attribute='value
            const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
            const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

            // loop through attribute/value pairs
            let match;
            while(match = text.match(regex2) || text.match(regex1)) {
                // get attribute and value from regex capture groups
                let attribute = match[1];
                let value = match[2];

                // remove from string
                text = text.substring(match.index + match[0].length);

                if (!attribute || !value)
                    break;

                // set attribute of parent element
                try {
                    element.setAttribute(attribute, value);
                } catch(error) {
                    console.log(error);
                }

                // special case for colspan
                if (attribute === 'colspan')
                    removeTableCells(element, value);
            }
        }

        // get list of comment elements in document
        function findComments() {
            const comments = [];

            // iterate over comment nodes in document
            function acceptNode(node) {
                return NodeFilter.FILTER_ACCEPT;
            }
            const iterator = document.createNodeIterator(
                document.body,
                NodeFilter.SHOW_COMMENT,
                acceptNode
            );
            let node;
            while(node = iterator.nextNode())
                comments.push(node);

            return comments;
        }

        // remove certain number of cells after specified cell
        function removeTableCells(cell, number) {
            number = parseInt(number);
            if (!number)
                return;

            // remove elements
            for(; number > 1; number--) {
                if (cell.nextElementSibling)
                    cell.nextElementSibling.remove();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- math plugin configuration -->

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        "CommonHTML": { linebreaks: { automatic: true } },
        "HTML-CSS": { linebreaks: { automatic: true } },
        "SVG": { linebreaks: { automatic: true } },
        "fast-preview": { disabled: true }
    });
</script>

<!-- math plugin -->

<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML'>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'MathJax' allows the proper rendering of
    // math/equations written in LaTeX.

    // https://www.mathjax.org/
</script>
<!-- annotations plugin configuration -->

<script>
    window.hypothesisConfig = function() {
        return {
            branding: {
                accentColor: '#2196f3',
                appBackgroundColor: '#f8f8f8',
                ctaBackgroundColor: '#f8f8f8',
                ctaTextColor: '#000000',
                selectionFontFamily: 'Open Sans, Helvetica, sans serif',
                annotationFontFamily: 'Open Sans, Helvetica, sans serif'
            }
        };
    };
</script>

<!-- annotations plugin -->

<script src='https://hypothes.is/embed.js'>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'Hypothesis' allows public annotation of the
    // manuscript.

    // https://web.hypothes.is/
</script>
<!-- analytics plugin -->

<!-- copy and paste code from Google Analytics or similar service here -->
</body>
</html>
